<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Wanxin Li">
<meta name="dcterms.date" content="2024-08-15">

<title>Shape Analysis of Cancer Cells – bioshape-analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-659650fc26dc25888fc1474f317bb8ac.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">bioshape-analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Shape Analysis of Cancer Cells</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">biology</div>
                <div class="quarto-category">bioinformatics</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Wanxin Li </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">August 15, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction-and-motivation" id="toc-introduction-and-motivation" class="nav-link active" data-scroll-target="#introduction-and-motivation">1. Introduction and Motivation</a></li>
  <li><a href="#dataset-description" id="toc-dataset-description" class="nav-link" data-scroll-target="#dataset-description">2. Dataset Description</a></li>
  <li><a href="#preprocessing" id="toc-preprocessing" class="nav-link" data-scroll-target="#preprocessing">3. Preprocessing</a>
  <ul class="collapse">
  <li><a href="#interpolation-encoding-discrete-curves-with-same-number-of-points" id="toc-interpolation-encoding-discrete-curves-with-same-number-of-points" class="nav-link" data-scroll-target="#interpolation-encoding-discrete-curves-with-same-number-of-points">Interpolation: Encoding Discrete Curves With Same Number of Points</a></li>
  <li><a href="#visualization-of-interpolated-dataset-of-curves" id="toc-visualization-of-interpolated-dataset-of-curves" class="nav-link" data-scroll-target="#visualization-of-interpolated-dataset-of-curves">Visualization of Interpolated Dataset of Curves</a></li>
  <li><a href="#remove-duplicate-sample-points-in-curves" id="toc-remove-duplicate-sample-points-in-curves" class="nav-link" data-scroll-target="#remove-duplicate-sample-points-in-curves">Remove duplicate sample points in curves</a></li>
  <li><a href="#alignment" id="toc-alignment" class="nav-link" data-scroll-target="#alignment">Alignment</a></li>
  </ul></li>
  <li><a href="#data-analysis" id="toc-data-analysis" class="nav-link" data-scroll-target="#data-analysis">4 Data Analysis</a>
  <ul class="collapse">
  <li><a href="#compute-mean-cell-shape-of-the-whole-dataset-global-mean-shape" id="toc-compute-mean-cell-shape-of-the-whole-dataset-global-mean-shape" class="nav-link" data-scroll-target="#compute-mean-cell-shape-of-the-whole-dataset-global-mean-shape">Compute Mean Cell Shape of the Whole Dataset: “Global” Mean Shape</a></li>
  </ul></li>
  <li><a href="#analyze-distances-to-the-global-mean-shape" id="toc-analyze-distances-to-the-global-mean-shape" class="nav-link" data-scroll-target="#analyze-distances-to-the-global-mean-shape">Analyze Distances to the “Global” Mean Shape</a></li>
  <li><a href="#visualization-of-the-mean-of-each-treatment" id="toc-visualization-of-the-mean-of-each-treatment" class="nav-link" data-scroll-target="#visualization-of-the-mean-of-each-treatment">Visualization of the Mean of each Treatment</a></li>
  <li><a href="#distance-of-the-cell-shapes-to-their-own-mean-shape" id="toc-distance-of-the-cell-shapes-to-their-own-mean-shape" class="nav-link" data-scroll-target="#distance-of-the-cell-shapes-to-their-own-mean-shape">Distance of the Cell Shapes to their Own Mean Shape</a></li>
  <li><a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction">Dimensionality Reduction</a></li>
  <li><a href="#multi-class-3-class-classification" id="toc-multi-class-3-class-classification" class="nav-link" data-scroll-target="#multi-class-3-class-classification">Multi-class (3-class) classification</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>This notebook is adapted from <a href="https://github.com/geomstats/geomstats/blob/main/notebooks/11_real_world_applications__cell_shapes_analysis.ipynb">this notebook</a> (Lead author: Nina Miolane).</p>
<p>This notebook studies <em>Osteosarcoma</em> (bone cancer) cells and the impact of drug treatment on their <em>morphological shapes</em>, by analyzing cell images obtained from fluorescence microscopy.</p>
<p>This analysis relies on the <em>elastic metric between discrete curves</em> from Geomstats. We will study to which extent this metric can detect how the cell shape is associated with the response to treatment.</p>
<p>The full papers analyzing this dataset are available at <span class="citation" data-cites="li2023using">Li et al. (<a href="#ref-li2023using" role="doc-biblioref">2023</a>)</span>, <span class="citation" data-cites="li2024unveiling">Li et al. (<a href="#ref-li2024unveiling" role="doc-biblioref">2024</a>)</span>.</p>
<center>
<img src="figs/illustration/cells_image.png" width="600px">
</center>
<p>Figure 1: Representative images of the cell lines using fluorescence microscopy, studied in this notebook (Image credit : Ashok Prasad). The cells nuclei (blue), the actin cytoskeleton (green) and the lipid membrane (red) of each cell are stained and colored. We only focus on the cell shape in our analysis.</p>
<section id="introduction-and-motivation" class="level1">
<h1>1. Introduction and Motivation</h1>
<p>Biological cells adopt a variety of shapes, determined by multiple processes and biophysical forces under the control of the cell. These shapes can be studied with different quantitative measures that reflect the cellular morphology <a href="#References">(MGCKCKDDRTWSBCC2018)</a>. With the emergence of large-scale biological cell image data, morphological studies have many applications. For example, measures of irregularity and spreading of cells allow accurate classification and discrimination between cancer cell lines treated with different drugs <a href="#References">(AXCFP2019)</a>.</p>
<p>As metrics defined on the shape space of curves, the <em>elastic metrics</em> <a href="#References">(SKJJ2010)</a> implemented in Geomstats are a potential tool for analyzing and comparing biological cell shapes. Their associated geodesics and geodesic distances provide a natural framework for optimally matching, deforming, and comparing cell shapes.</p>
<div id="46493d28" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> decimal <span class="im">import</span> Decimal</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geomstats.backend <span class="im">as</span> gs</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> common <span class="im">import</span> <span class="op">*</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> stats</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn <span class="im">import</span> manifold</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>gs.random.seed(<span class="dv">2021</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4cae6b09" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>base_path <span class="op">=</span> <span class="st">"/home/wanxinli/dyn/dyn/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data_path <span class="op">=</span> os.path.join(base_path, <span class="st">"datasets"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>dataset_name <span class="op">=</span> <span class="st">'osteosarcoma'</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>figs_dir <span class="op">=</span> os.path.join(<span class="st">"/home/wanxinli/dyn/dyn/figs"</span>, dataset_name)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>savefig <span class="op">=</span> <span class="va">False</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># If compute for the first time, we need to compute pairwise distances and run DeCOr-MDS</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Otherwise, we can just use the pre-computed results</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>first_time <span class="op">=</span> <span class="va">False</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Will save figs to </span><span class="sc">{</span>figs_dir<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dataset-description" class="level1">
<h1>2. Dataset Description</h1>
<p>We study a dataset of mouse <em>Osteosarcoma</em> imaged cells <a href="#References">(AXCFP2019)</a>. The dataset contains two different cancer cell lines : <em>DLM8</em> and <em>DUNN</em>, respectively representing a more agressive and a less agressive cancer. Among these cells, some have also been treated with different single drugs that perturb the cellular cytoskeleton. Overall, we can label each cell according to their cell line (<em>DLM8</em> and <em>DUNN</em>), and also if it is a <em>control</em> cell (no treatment), or has been treated with one of the following drugs : <em>Jasp</em> (jasplakinolide) and <em>Cytd</em> (cytochalasin D).</p>
<p>Each cell comes from a raw image containing a set of cells, which was thresholded to generate binarized images.</p>

<img src="figs/illustration/binarized_cells.png" width="400px">

<p>After binarizing the images, contouring was used to isolate each cell, and to extract their boundaries as a counter-clockwise ordered list of 2D coordinates, which corresponds to the representation of discrete curve in Geomstats. We load these discrete curves into the notebook.</p>
<div id="ca9a24ea" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> geomstats.datasets.utils <span class="im">as</span> data_utils</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>cells, lines, treatments <span class="op">=</span> data_utils.load_cells()</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Total number of cells : </span><span class="sc">{</span><span class="bu">len</span>(cells)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of cells : 650</code></pre>
</div>
</div>
<p>The cells are grouped by treatment class in the dataset : - the <em>control</em> cells, - the cells treated with <em>Cytd</em>, - and the ones treated with <em>Jasp</em>.</p>
<p>Additionally, in each of these classes, there are two cell lines : - the <em>DLM8</em> cells, and - the <em>DUNN</em> ones.</p>
<p>Before using the dataset, we check for duplicates in the dataset.</p>
<p>We compute the pairwise distance between two cells. If the pairwise distance is smaller than 0.1, we visualize the corresponding cells to check they are duplicates.</p>
<div id="20de2254" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>tol <span class="op">=</span> <span class="fl">1e-1</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, cell_i <span class="kw">in</span> <span class="bu">enumerate</span>(cells):</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j, cell_j <span class="kw">in</span> <span class="bu">enumerate</span>(cells):</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> i <span class="op">!=</span> j <span class="kw">and</span> cell_i.shape[<span class="dv">0</span>] <span class="op">==</span> cell_j.shape[<span class="dv">0</span>]:</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>            dist <span class="op">=</span> np.<span class="bu">sum</span>(np.sqrt(np.<span class="bu">sum</span>((cell_i<span class="op">-</span>cell_j)<span class="op">**</span><span class="dv">2</span>,axis<span class="op">=</span><span class="dv">1</span>)))</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> dist <span class="op">&lt;</span> tol:</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"cell indices are: </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>j<span class="sc">}</span><span class="ss">, </span><span class="sc">{</span>lines[i]<span class="sc">}</span><span class="ss">, </span><span class="sc">{</span>lines[j]<span class="sc">}</span><span class="ss">, </span><span class="sc">{</span>treatments[i]<span class="sc">}</span><span class="ss">, </span><span class="sc">{</span>treatments[j]<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cell indices are: 363 and 396, dlm8, dlm8, cytd, cytd
cell indices are: 396 and 363, dlm8, dlm8, cytd, cytd
cell indices are: 513 and 519, dlm8, dlm8, jasp, jasp
cell indices are: 519 and 513, dlm8, dlm8, jasp, jasp</code></pre>
</div>
</div>
<div id="635fb854" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>pair_indices <span class="op">=</span> [<span class="dv">363</span>, <span class="dv">396</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">5</span>))</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">121</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>index_0 <span class="op">=</span> pair_indices[<span class="dv">0</span>]</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>plt.scatter(cells[index_0][:, <span class="dv">0</span>], cells[index_0][:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Cell </span><span class="sc">{</span>index_0<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">122</span>)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>index_1 <span class="op">=</span> pair_indices[<span class="dv">1</span>]</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>plt.scatter(cells[index_1][:, <span class="dv">0</span>], cells[index_1][:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Cell </span><span class="sc">{</span>index_1<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>Text(0.5, 1.0, 'Cell 396')</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-6-output-2.png" width="809" height="431" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="92046d87" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>pair_indices <span class="op">=</span> [<span class="dv">513</span>, <span class="dv">519</span>]</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">5</span>))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">121</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>index_0 <span class="op">=</span> pair_indices[<span class="dv">0</span>]</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>plt.scatter(cells[index_0][:, <span class="dv">0</span>], cells[index_0][:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Cell </span><span class="sc">{</span>index_0<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">122</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>index_1 <span class="op">=</span> pair_indices[<span class="dv">1</span>]</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>plt.scatter(cells[index_1][:, <span class="dv">0</span>], cells[index_1][:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Cell </span><span class="sc">{</span>index_1<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>Text(0.5, 1.0, 'Cell 519')</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-7-output-2.png" width="809" height="431" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Check the category indices in order to remove corresponding cells in <code>ds_align</code></p>
<div id="ed589fb9" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>delete_indices <span class="op">=</span> [<span class="dv">363</span>, <span class="dv">396</span>, <span class="dv">513</span>, <span class="dv">519</span>]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>category_count <span class="op">=</span> {}</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>global_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(cells)):</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    treatment <span class="op">=</span> treatments[i]</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    line <span class="op">=</span> lines[i]</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> treatment <span class="kw">not</span> <span class="kw">in</span> category_count:</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>        category_count[treatment] <span class="op">=</span> {}</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> line <span class="kw">not</span> <span class="kw">in</span> category_count[treatment]:</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>        category_count[treatment][line] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if global_count in delete_indices:</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     print(treatment, line, category_count[treatment][line])</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    category_count[treatment][line] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>    global_count <span class="op">+=</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since 363th, 396th and 513th, 519th are duplicates of each other and after visualization we see they are poor quality cells with overlapping adjacent cells, we remove them from our dataset.</p>
<div id="e00212aa" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> remove_cells(cells, lines, treatments, delete_indices):</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">""" </span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Remove cells of control group from cells, lines and treatments</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co">    :param list[int] delete_indices: the indices to delete</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    delete_indices <span class="op">=</span> <span class="bu">sorted</span>(delete_indices, reverse<span class="op">=</span><span class="va">True</span>) <span class="co"># to prevent change in index when deleting elements</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Delete elements</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    cells <span class="op">=</span> del_arr_elements(cells, delete_indices)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> <span class="bu">list</span>(np.delete(np.array(lines), delete_indices, axis<span class="op">=</span><span class="dv">0</span>))</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    treatments <span class="op">=</span> <span class="bu">list</span>(np.delete(np.array(treatments), delete_indices, axis<span class="op">=</span><span class="dv">0</span>))</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> cells, lines, treatments</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ff66db3d" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>delete_indices <span class="op">=</span> [<span class="dv">363</span>, <span class="dv">396</span>, <span class="dv">513</span>, <span class="dv">519</span>]</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>cells, lines, treatments <span class="op">=</span> remove_cells(cells, lines, treatments, delete_indices)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># print(len(cells), len(lines), len(treatments))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is shown by displaying the unique elements in the lists <code>treatments</code> and <code>lines</code>:</p>
<div id="37d8e59a" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>TREATMENTS <span class="op">=</span> gs.unique(treatments)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(TREATMENTS)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>LINES <span class="op">=</span> gs.unique(lines)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(LINES)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>METRICS <span class="op">=</span> [<span class="st">'SRV'</span>, <span class="st">'Linear'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['control' 'cytd' 'jasp']
['dlm8' 'dunn']</code></pre>
</div>
</div>
<p>The size of each class is displayed below:</p>
<div id="fb2628df" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>ds <span class="op">=</span> {}</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>n_cells_arr <span class="op">=</span> gs.zeros((<span class="dv">3</span>, <span class="dv">2</span>))</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss"> :"</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    ds[treatment] <span class="op">=</span> {}</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j, line <span class="kw">in</span> <span class="bu">enumerate</span>(LINES):</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>        to_keep <span class="op">=</span> gs.array(</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>            [</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>                one_treatment <span class="op">==</span> treatment <span class="kw">and</span> one_line <span class="op">==</span> line</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>                <span class="cf">for</span> one_treatment, one_line <span class="kw">in</span> <span class="bu">zip</span>(treatments, lines)</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>            ]</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>        ds[treatment][line] <span class="op">=</span> [</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>            cell_i <span class="cf">for</span> cell_i, to_keep_i <span class="kw">in</span> <span class="bu">zip</span>(cells, to_keep) <span class="cf">if</span> to_keep_i</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>        ]</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>        nb <span class="op">=</span> <span class="bu">len</span>(ds[treatment][line])</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="ss"> </span><span class="sc">{</span>nb<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>        n_cells_arr[i, j] <span class="op">=</span> nb</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>n_cells_df <span class="op">=</span> pd.DataFrame({<span class="st">"dlm8"</span>: n_cells_arr[:, <span class="dv">0</span>], <span class="st">"dunn"</span>: n_cells_arr[:, <span class="dv">1</span>]})</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>n_cells_df <span class="op">=</span> n_cells_df.set_index(TREATMENTS)</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>display(n_cells_df)</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a><span class="co"># display(ds)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>control :
     114 dlm8
     204 dunn
cytd :
     80 dlm8
     93 dunn
jasp :
     60 dlm8
     95 dunn</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">dlm8</th>
<th data-quarto-table-cell-role="th">dunn</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">control</td>
<td>114.0</td>
<td>204.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">cytd</td>
<td>80.0</td>
<td>93.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">jasp</td>
<td>60.0</td>
<td>95.0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>We have organized the cell data into the dictionnary <code>ds</code>. Before proceeding to the actual data analysis, we provide an auxiliary function <code>apply_func_to_ds</code>.</p>
<div id="efd38e73" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> apply_func_to_ds(input_ds, func):</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Apply the input function func to the input dictionnary input_ds.</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">    This function goes through the dictionnary structure and applies</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">    func to every cell in input_ds[treatment][line].</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">    It stores the result in a dictionnary output_ds that is returned</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co">    to the user.</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">    Parameters</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">    ----------</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co">    input_ds : dict</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="co">        Input dictionnary, with keys treatment-line.</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="co">    func : callable</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="co">        Function to be applied to the values of the dictionnary, i.e.</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="co">        the cells.</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="co">    -------</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="co">    output_ds : dict</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="co">        Output dictionnary, with the same keys as input_ds.</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>    output_ds <span class="op">=</span> {}</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>        output_ds[treatment] <span class="op">=</span> {}</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>            output_list <span class="op">=</span> []</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> one_cell <span class="kw">in</span> input_ds[treatment][line]:</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>                output_list.append(func(one_cell))</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>            output_ds[treatment][line] <span class="op">=</span> gs.array(output_list)</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> output_ds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can move on to the actual data analysis, starting with a preprocessing of the cell boundaries.</p>
</section>
<section id="preprocessing" class="level1">
<h1>3. Preprocessing</h1>
<section id="interpolation-encoding-discrete-curves-with-same-number-of-points" class="level3">
<h3 class="anchored" data-anchor-id="interpolation-encoding-discrete-curves-with-same-number-of-points">Interpolation: Encoding Discrete Curves With Same Number of Points</h3>
<p>As we need discrete curves with the same number of sampled points to compute pairwise distances, the following interpolation is applied to each curve, after setting the number of sampling points.</p>
<p>To set up the number of sampling points, you can edit the following line in the next cell:</p>
<div id="b621d526" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> interpolate(curve, nb_points):</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Interpolate a discrete curve with nb_points from a discrete curve.</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">    -------</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co">    interpolation : discrete curve with nb_points points</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    old_length <span class="op">=</span> curve.shape[<span class="dv">0</span>]</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>    interpolation <span class="op">=</span> gs.zeros((nb_points, <span class="dv">2</span>))</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    incr <span class="op">=</span> old_length <span class="op">/</span> nb_points</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    pos <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(nb_points):</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>        index <span class="op">=</span> <span class="bu">int</span>(gs.floor(pos))</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>        interpolation[i] <span class="op">=</span> curve[index] <span class="op">+</span> (pos <span class="op">-</span> index) <span class="op">*</span> (</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>            curve[(index <span class="op">+</span> <span class="dv">1</span>) <span class="op">%</span> old_length] <span class="op">-</span> curve[index]</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>        pos <span class="op">+=</span> incr</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> interpolation</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>k_sampling_points <span class="op">=</span> <span class="dv">2000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To illustrate the result of this interpolation, we compare for a randomly chosen cell the original curve with the correponding interpolated one (to visualize another cell, you can simply re-run the code).</p>
<div id="652ed9a0" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>cell_rand <span class="op">=</span> cells[index]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>cell_interpolation <span class="op">=</span> interpolate(cell_rand, k_sampling_points)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">5</span>))</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">121</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>plt.scatter(cell_rand[:, <span class="dv">0</span>], cell_rand[:, <span class="dv">1</span>], color<span class="op">=</span><span class="st">'black'</span>, s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>plt.plot(cell_rand[:, <span class="dv">0</span>], cell_rand[:, <span class="dv">1</span>])</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Original curve (</span><span class="sc">{</span><span class="bu">len</span>(cell_rand)<span class="sc">}</span><span class="ss"> points)"</span>)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"off"</span>)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">122</span>)</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>plt.scatter(cell_interpolation[:, <span class="dv">0</span>], cell_interpolation[:, <span class="dv">1</span>], color<span class="op">=</span><span class="st">'black'</span>, s<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>plt.plot(cell_interpolation[:, <span class="dv">0</span>], cell_interpolation[:, <span class="dv">1</span>])</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f"Interpolated curve (</span><span class="sc">{</span>k_sampling_points<span class="sc">}</span><span class="ss"> points)"</span>)</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"off"</span>)</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"interpolation.svg"</span>))</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"interpolation.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-15-output-1.png" width="1135" height="409" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As the interpolation is working as expected, we use the auxiliary function <code>apply_func_to_ds</code> to apply the function <code>func=interpolate</code> to the dataset <code>ds</code>, i.e.&nbsp;the dictionnary containing the cells boundaries.</p>
<p>We obtain a new dictionnary, <code>ds_interp</code>, with the interpolated cell boundaries.</p>
<div id="870f3c09" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ds_interp <span class="op">=</span> apply_func_to_ds(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    input_ds<span class="op">=</span>ds, func<span class="op">=</span><span class="kw">lambda</span> x: interpolate(x, k_sampling_points)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The shape of an array of cells in <code>ds_interp[treatment][cell]</code> is therefore: <code>("number of cells in treatment-line", "number of sampling points", 2)</code>, where 2 refers to the fact that we are considering cell shapes in 2D.</p>
</section>
<section id="visualization-of-interpolated-dataset-of-curves" class="level3">
<h3 class="anchored" data-anchor-id="visualization-of-interpolated-dataset-of-curves">Visualization of Interpolated Dataset of Curves</h3>
<p>We visualize the curves obtained, for a sample of control cells and treated cells (top row shows control, i.e.&nbsp;non-treated cells; bottom rows shows treated cells) across cell lines (left and blue for dlm8 and right and orange for dunn).</p>
<div id="35c6f52a" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>n_cells_to_plot <span class="op">=</span> <span class="dv">5</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># radius = 800</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">16</span>, <span class="dv">6</span>))</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>count <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>        cell_data <span class="op">=</span> ds_interp[treatment][line]</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i_to_plot <span class="kw">in</span> <span class="bu">range</span>(n_cells_to_plot):</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>            cell <span class="op">=</span> gs.random.choice(cell_data)</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>            fig.add_subplot(<span class="dv">3</span>, <span class="dv">2</span> <span class="op">*</span> n_cells_to_plot, count)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>            count <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>            plt.plot(cell[:, <span class="dv">0</span>], cell[:, <span class="dv">1</span>], color<span class="op">=</span><span class="st">"C"</span> <span class="op">+</span> <span class="bu">str</span>(i))</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>            <span class="co"># plt.xlim(-radius, radius)</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>            <span class="co"># plt.ylim(-radius, radius)</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>            plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>            plt.axis(<span class="st">"off"</span>)</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i_to_plot <span class="op">==</span> n_cells_to_plot <span class="op">//</span> <span class="dv">2</span>:</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>                plt.title(<span class="ss">f"</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss">   -   </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"sample_cells.svg"</span>))</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"sample_cells.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-17-output-1.png" width="1210" height="491" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Visual inspection of these curves seems to indicate more protusions appearing in treated cells, compared with control ones. This is in agreement with the physiological impact of the drugs, which are known to perturb the internal cytoskeleton connected to the cell membrane. Using the elastic metric, our goal will be to see if we can quantitatively confirm these differences.</p>
</section>
<section id="remove-duplicate-sample-points-in-curves" class="level3">
<h3 class="anchored" data-anchor-id="remove-duplicate-sample-points-in-curves">Remove duplicate sample points in curves</h3>
<p>During interpolation it is likely that some of the discrete curves in the dataset are downsampled from higher number of discrete data points to lower number of data points. Hence, two sampled data points that are close enough may end up overlapping after interpolation and hence such data points have to be dealt with specifically.</p>
<div id="965e14a2" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> preprocess(curve, tol<span class="op">=</span><span class="fl">1e-10</span>):</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Preprocess curve to ensure that there are no consecutive duplicate points.</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co">    -------</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co">    curve : discrete curve</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    dist <span class="op">=</span> curve[<span class="dv">1</span>:] <span class="op">-</span> curve[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    dist_norm <span class="op">=</span> np.sqrt(np.<span class="bu">sum</span>(np.square(dist), axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> np.<span class="bu">any</span>( dist_norm <span class="op">&lt;</span> tol ):</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(curve)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> np.sqrt(np.<span class="bu">sum</span>(np.square(curve[i<span class="op">+</span><span class="dv">1</span>] <span class="op">-</span> curve[i]), axis<span class="op">=</span><span class="dv">0</span>)) <span class="op">&lt;</span> tol:</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>                curve[i<span class="op">+</span><span class="dv">1</span>] <span class="op">=</span> (curve[i] <span class="op">+</span> curve[i<span class="op">+</span><span class="dv">2</span>]) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> curve</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="2280bbd4" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ds_proc <span class="op">=</span> apply_func_to_ds(ds_interp, func<span class="op">=</span><span class="kw">lambda</span> x: preprocess(x))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check we did not loss any cells after duplicates</p>
<div id="21b1c2f6" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss"> using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span><span class="bu">len</span>(ds_proc[treatment][line])<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>control and dlm8 using SRV: 114
control and dlm8 using Linear: 114
control and dunn using SRV: 204
control and dunn using Linear: 204
cytd and dlm8 using SRV: 80
cytd and dlm8 using Linear: 80
cytd and dunn using SRV: 93
cytd and dunn using Linear: 93
jasp and dlm8 using SRV: 60
jasp and dlm8 using Linear: 60
jasp and dunn using SRV: 95
jasp and dunn using Linear: 95</code></pre>
</div>
</div>
</section>
<section id="alignment" class="level3">
<h3 class="anchored" data-anchor-id="alignment">Alignment</h3>
<p>Our goal is to study the cell boundaries in our dataset, as points in a shape space of closed curves quotiented by translation, scaling, and rotation, so these transformations do not affect our measure of distance between curves.</p>
<p>In practice, we apply functions that were initially designed to center (substract the barycenter), rescale (divide by the Frobenius norm) and then reparameterize (only for SRV metric).</p>
<p>Since the alignment procedure takes 30 minutes, we ran <code>osteosarocoma_align.py</code> and saved the results in <code>~/dyn/datasets/osteosarcoma/aligned</code></p>
<p>Load aligned cells from txt files. These files were generated by calling <code>align</code> function in <code>common.py</code>.</p>
<p>We get the aligned cells from preprocessed dataset.</p>
<p>Furthermore, we align the barycenters of the cells to the barycenter of the projected base curve, and (optionally) flip the cell.</p>
<div id="a7d2a4be" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> align_barycenter(cell, centroid_x, centroid_y, flip):</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">""" </span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Align the the barycenter of the cell to ref centeriod and flip the cell against the xaxis of the centriod if flip is True. </span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co">    :param 2D np array cell: cell to align</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co">    :param float centroid_x: the x coordinates of the projected BASE_CURVE</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co">    :param float centroid_y: the y coordinates of the projected BASE_CURVE</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="co">    :param bool flip: flip the cell against x = centroid x if True </span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    cell_bc <span class="op">=</span> np.mean(cell, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    aligned_cell <span class="op">=</span> cell<span class="op">+</span>[centroid_x, centroid_y]<span class="op">-</span>cell_bc</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> flip:</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>        aligned_cell[:, <span class="dv">0</span>] <span class="op">=</span> <span class="dv">2</span><span class="op">*</span>centroid_x<span class="op">-</span>aligned_cell[:, <span class="dv">0</span>]</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Flip the order of the points</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>        med_index <span class="op">=</span> <span class="bu">int</span>(np.floor(aligned_cell.shape[<span class="dv">0</span>]<span class="op">/</span><span class="dv">2</span>))</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>        flipped_aligned_cell <span class="op">=</span> np.concatenate((aligned_cell[med_index:], aligned_cell[:med_index]), axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>        flipped_aligned_cell <span class="op">=</span> np.flipud(flipped_aligned_cell)</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>        aligned_cell <span class="op">=</span> flipped_aligned_cell</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> aligned_cell</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_centroid(base_curve):</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>    total_space <span class="op">=</span> DiscreteCurvesStartingAtOrigin(k_sampling_points<span class="op">=</span>k_sampling_points)</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>    proj_base_curve <span class="op">=</span> total_space.projection(base_curve)</span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>    base_centroid <span class="op">=</span> np.mean(proj_base_curve, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> base_centroid[<span class="dv">0</span>], base_centroid[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="864040fa" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>delete_indices <span class="op">=</span> [<span class="dv">363</span>, <span class="dv">396</span>, <span class="dv">513</span>, <span class="dv">519</span>]</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>aligned_base_folder <span class="op">=</span> os.path.join(data_path, dataset_name, <span class="st">"aligned"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>BASE_CURVE <span class="op">=</span> generate_ellipse(k_sampling_points)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>centroid_x, centroid_y <span class="op">=</span> get_centroid(BASE_CURVE)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>ds_align <span class="op">=</span> {}</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    ds_align[metric] <span class="op">=</span> {}</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> metric <span class="op">==</span> <span class="st">'SRV'</span>:</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>        aligned_folder <span class="op">=</span> os.path.join(aligned_base_folder, <span class="st">'projection_rescale_rotation_reparameterization'</span>)</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> metric <span class="op">==</span> <span class="st">'Linear'</span>:</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>        aligned_folder <span class="op">=</span> os.path.join(aligned_base_folder, <span class="st">'projection_rescale_rotation_reparameterization'</span>)</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>        ds_align[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>            ds_align[metric][treatment][line] <span class="op">=</span> []</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>            cell_num <span class="op">=</span> <span class="bu">len</span>(ds_proc[treatment][line])</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> line <span class="op">==</span> <span class="st">'dlm8'</span> <span class="kw">and</span> (treatment <span class="op">==</span> <span class="st">'cytd'</span> <span class="kw">or</span> treatment <span class="op">==</span> <span class="st">'jasp'</span>):</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>                cell_num <span class="op">+=</span> <span class="dv">2</span></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(cell_num):</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Do not load duplicate cells</span></span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>                <span class="co"># cytd dlm8 45</span></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>                <span class="co"># cytd dlm8 78</span></span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>                <span class="co"># jasp dlm8 20</span></span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>                <span class="co"># jasp dlm8 26</span></span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> (treatment <span class="op">==</span> <span class="st">'cytd'</span> <span class="kw">and</span> line <span class="op">==</span> <span class="st">'dlm8'</span> <span class="kw">and</span> (i <span class="op">==</span> <span class="dv">45</span> <span class="kw">or</span> i <span class="op">==</span> <span class="dv">78</span>)) <span class="kw">or</span> <span class="op">\</span></span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>                    (treatment <span class="op">==</span> <span class="st">'jasp'</span> <span class="kw">and</span> line <span class="op">==</span> <span class="st">'dlm8'</span> <span class="kw">and</span> (i <span class="op">==</span> <span class="dv">20</span> <span class="kw">or</span> i <span class="op">==</span> <span class="dv">26</span>)):</span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">continue</span></span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>                file_path <span class="op">=</span> os.path.join(aligned_folder, <span class="ss">f"</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">.txt"</span>)</span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> os.path.exists(file_path):</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>                    cell <span class="op">=</span> np.loadtxt(file_path)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>                    ds_align[metric][treatment][line].append(cell)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check we did not loss any cells after alignment</p>
<div id="00be86a0" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss"> using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span><span class="bu">len</span>(ds_align[metric][treatment][line])<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>control and dlm8 using SRV: 113
control and dlm8 using Linear: 113
control and dunn using SRV: 199
control and dunn using Linear: 199
cytd and dlm8 using SRV: 74
cytd and dlm8 using Linear: 74
cytd and dunn using SRV: 92
cytd and dunn using Linear: 92
jasp and dlm8 using SRV: 56
jasp and dlm8 using Linear: 56
jasp and dunn using SRV: 91
jasp and dunn using Linear: 91</code></pre>
</div>
</div>
<p>Update <code>lines</code> and <code>treatments</code></p>
<div id="8750181d" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>treatments <span class="op">=</span> []</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>lines <span class="op">=</span> []</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>        treatments.extend([treatment]<span class="op">*</span><span class="bu">len</span>(ds_align[<span class="st">'SRV'</span>][treatment][line]))</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>        lines.extend([line]<span class="op">*</span><span class="bu">len</span>(ds_align[<span class="st">'SRV'</span>][treatment][line]))</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>treatments <span class="op">=</span> np.array(treatments)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>lines <span class="op">=</span> np.array(lines)</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"treatment length is:"</span>, <span class="bu">len</span>(treatments), <span class="st">"lines length is:"</span>, <span class="bu">len</span>(lines))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>treatment length is: 625 lines length is: 625</code></pre>
</div>
</div>
<p>Visualize reference cell, unaligned cell and aligned cell.</p>
<div id="56fbb8c1" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>unaligned_cell <span class="op">=</span> ds_proc[<span class="st">"control"</span>][<span class="st">"dlm8"</span>][index]</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>aligned_cell <span class="op">=</span> ds_align[metric][<span class="st">"control"</span>][<span class="st">"dlm8"</span>][index]</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>first_round_aligned_folder <span class="op">=</span> os.path.join(aligned_base_folder, <span class="st">'projection_rescale_rotation_reparameterization_first_round'</span>)</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>reference_path <span class="op">=</span> os.path.join(first_round_aligned_folder, <span class="ss">f"reference.txt"</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>mean_first_round <span class="op">=</span> np.loadtxt(reference_path)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">5</span>))</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">131</span>)</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>plt.plot(mean_first_round[:, <span class="dv">0</span>], mean_first_round[:, <span class="dv">1</span>])</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>plt.plot([mean_first_round[<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>], mean_first_round[<span class="dv">0</span>, <span class="dv">0</span>]], [mean_first_round[<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>], mean_first_round[<span class="dv">0</span>, <span class="dv">1</span>]], <span class="st">'tab:blue'</span>)</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>plt.scatter(mean_first_round[:, <span class="dv">0</span>], mean_first_round[:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>, c<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>plt.plot(mean_first_round[<span class="dv">0</span>, <span class="dv">0</span>], mean_first_round[<span class="dv">0</span>, <span class="dv">1</span>], <span class="st">"ro"</span>)</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Reference curve"</span>)</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">132</span>)</span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>plt.plot(unaligned_cell[:, <span class="dv">0</span>], unaligned_cell[:, <span class="dv">1</span>])</span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>plt.scatter(unaligned_cell[:, <span class="dv">0</span>], unaligned_cell[:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>, c<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>plt.plot(unaligned_cell[<span class="dv">0</span>, <span class="dv">0</span>], unaligned_cell[<span class="dv">0</span>, <span class="dv">1</span>], <span class="st">"ro"</span>)</span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Unaligned curve"</span>)</span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">133</span>)</span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a>plt.plot(aligned_cell[:, <span class="dv">0</span>], aligned_cell[:, <span class="dv">1</span>])</span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a>plt.scatter(aligned_cell[:, <span class="dv">0</span>], aligned_cell[:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">4</span>, c<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a>plt.plot(aligned_cell[<span class="dv">0</span>, <span class="dv">0</span>], aligned_cell[<span class="dv">0</span>, <span class="dv">1</span>], <span class="st">"ro"</span>)</span>
<span id="cb33-31"><a href="#cb33-31" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb33-32"><a href="#cb33-32" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Aligned curve"</span>)</span>
<span id="cb33-33"><a href="#cb33-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-34"><a href="#cb33-34" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb33-35"><a href="#cb33-35" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"alignment.svg"</span>))</span>
<span id="cb33-36"><a href="#cb33-36" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"alignment.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-25-output-1.png" width="1185" height="431" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>In the plot above, the red dot shows the start of the parametrization of each curve. The right curve has been rotated from the curve in the middle, to be aligned with the left (reference) curve, which represents the first cell of the dataset. The starting point (in red) of this right curve has been also set to align with the reference.</p>
</section>
</section>
<section id="data-analysis" class="level1">
<h1>4 Data Analysis</h1>
<section id="compute-mean-cell-shape-of-the-whole-dataset-global-mean-shape" class="level2">
<h2 class="anchored" data-anchor-id="compute-mean-cell-shape-of-the-whole-dataset-global-mean-shape">Compute Mean Cell Shape of the Whole Dataset: “Global” Mean Shape</h2>
<p>We want to compute the mean cell shape of the whole dataset. Thus, we first combine all the cell shape data into a single array.</p>
<div id="7017b4bf" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>CURVES_SPACE_SRV <span class="op">=</span> DiscreteCurvesStartingAtOrigin(ambient_dim<span class="op">=</span><span class="dv">2</span>, k_sampling_points<span class="op">=</span>k_sampling_points)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="893a7962" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>cell_shapes_list <span class="op">=</span> {}</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    cell_shapes_list[metric] <span class="op">=</span> []</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>            cell_shapes_list[metric].extend(ds_align[metric][treatment][line])</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>cell_shapes <span class="op">=</span> {}</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    cell_shapes[metric] <span class="op">=</span> gs.array(cell_shapes_list[metric])</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cell_shapes[<span class="st">'SRV'</span>].shape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(625, 1999, 2)</code></pre>
</div>
</div>
<p>Remove outliers using DeCOr-MDS, together for DUNN and DLM8 cell lines.</p>
<div id="668e0504" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> linear_dist(cell1, cell2):</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> gs.linalg.norm(cell1 <span class="op">-</span> cell2)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> srv_dist(cell1, cell2):</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    CURVES_SPACE_SRV.equip_with_metric(SRVMetric)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> CURVES_SPACE_SRV.metric.dist(cell1, cell2)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="co"># compute pairwise distances, we only need to compute it once and save the results </span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>pairwise_dists <span class="op">=</span> {}</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> first_time:</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>    metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>    pairwise_dists[metric] <span class="op">=</span> parallel_dist(cell_shapes[metric], srv_dist, k_sampling_points)</span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>    metric <span class="op">=</span> <span class="st">'Linear'</span> </span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>    pairwise_dists[metric] <span class="op">=</span> parallel_dist(cell_shapes[metric], linear_dist, k_sampling_points)</span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a>        np.savetxt(os.path.join(data_path, dataset_name, <span class="st">"distance_matrix"</span>, <span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">_matrix.txt"</span>), pairwise_dists[metric])</span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a>        pairwise_dists[metric] <span class="op">=</span> np.loadtxt(os.path.join(data_path, dataset_name, <span class="st">"distance_matrix"</span>, <span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">_matrix.txt"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="0f664c1a" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># to remove 132 and 199</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>one_cell <span class="op">=</span> cell_shapes[<span class="st">'Linear'</span>][<span class="dv">199</span>]</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>plt.plot(one_cell[:, <span class="dv">0</span>], one_cell[:, <span class="dv">1</span>], c<span class="op">=</span><span class="ss">f"gray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-29-output-1.png" width="599" height="411" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="89d5f06f" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run DeCOr-MDS</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>dim_start <span class="op">=</span> <span class="dv">2</span> <span class="co"># we know the subspace dimension is 3, we set start and end to 3 to reduce runtime </span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>dim_end <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="co"># dim_start = 3</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="co"># dim_end = 3</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>std_multi <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> first_time:</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>    subspace_dim, outlier_indices <span class="op">=</span> find_subspace_dim(pairwise_dists[metric], dim_start, dim_end, std_multi)</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"subspace dimension is: </span><span class="sc">{</span>subspace_dim<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"outlier_indices are: </span><span class="sc">{</span>outlier_indices<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Visualize outlier cells to see if they are artifacts</p>
<div id="a6340654" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> first_time:</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>    fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>        nrows<span class="op">=</span> <span class="dv">1</span>,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>        ncols<span class="op">=</span><span class="bu">len</span>(outlier_indices),</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>        figsize<span class="op">=</span>(<span class="dv">2</span><span class="op">*</span><span class="bu">len</span>(outlier_indices), <span class="dv">2</span>),</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, outlier_index <span class="kw">in</span> <span class="bu">enumerate</span>(outlier_indices):</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>        one_cell <span class="op">=</span> cell_shapes[metric][outlier_index]</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>        ax <span class="op">=</span> axes[i]</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>        ax.plot(one_cell[:, <span class="dv">0</span>], one_cell[:, <span class="dv">1</span>], c<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>j<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>        ax.set_title(<span class="ss">f"</span><span class="sc">{</span>outlier_index<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Turn off tick labels</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>        ax.set_yticklabels([])</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>        ax.set_xticklabels([])</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>        ax.set_xticks([])</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>        ax.set_yticks([])</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>        ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>        ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>        ax.spines[<span class="st">"bottom"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>        ax.spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>    plt.suptitle(<span class="ss">f""</span>, y<span class="op">=-</span><span class="fl">0.01</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plt.savefig(os.path.join(figs_dir, "outlier.svg"))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e9426198" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>delete_indices <span class="op">=</span> [<span class="dv">132</span>, <span class="dv">199</span>]</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    nrows<span class="op">=</span> <span class="dv">1</span>,</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="bu">len</span>(delete_indices),</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">2</span><span class="op">*</span><span class="bu">len</span>(delete_indices), <span class="dv">2</span>),</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, outlier_index <span class="kw">in</span> <span class="bu">enumerate</span>(delete_indices):</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>    one_cell <span class="op">=</span> cell_shapes[metric][outlier_index]</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> axes[i]</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>    ax.plot(one_cell[:, <span class="dv">0</span>], one_cell[:, <span class="dv">1</span>], c<span class="op">=</span><span class="ss">f"gray"</span>)</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="ss">f"</span><span class="sc">{</span>outlier_index<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.axis("off")</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Turn off tick labels</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>    ax.set_yticklabels([])</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>    ax.set_xticklabels([])</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks([])</span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks([])</span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"bottom"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="ss">f""</span>, y<span class="op">=-</span><span class="fl">0.01</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"delete_outlier.svg"</span>))</span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"delete_outlier.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-32-output-1.png" width="374" height="182" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>After visual inspection, we decide to remove the outlier cells</p>
<div id="cac38e45" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> remove_ds_two_layer(ds, delete_indices):</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>    global_i <span class="op">=</span> <span class="bu">sum</span>(<span class="bu">len</span>(v) <span class="cf">for</span> values <span class="kw">in</span> ds.values() <span class="cf">for</span> v <span class="kw">in</span> values.values())<span class="op">-</span><span class="dv">1</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> treatment <span class="kw">in</span> <span class="bu">reversed</span>(<span class="bu">list</span>(ds.keys())):</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>        treatment_values <span class="op">=</span> ds[treatment]</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> <span class="bu">reversed</span>(<span class="bu">list</span>(treatment_values.keys())):</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>            line_cells <span class="op">=</span> treatment_values[line]</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> i, _ <span class="kw">in</span> <span class="bu">reversed</span>(<span class="bu">list</span>(<span class="bu">enumerate</span>(line_cells))):</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> global_i <span class="kw">in</span> delete_indices:</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>                    <span class="bu">print</span>(np.array(ds[treatment][line][:i]).shape, np.array(ds[treatment][line][i<span class="op">+</span><span class="dv">1</span>:]).shape)</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="bu">len</span>(np.array(ds[treatment][line][:i]).shape) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>                        ds[treatment][line] <span class="op">=</span> np.array(ds[treatment][line][i<span class="op">+</span><span class="dv">1</span>:])</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">elif</span> <span class="bu">len</span>(np.array(ds[treatment][line][i<span class="op">+</span><span class="dv">1</span>:]).shape) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>                        ds[treatment][line] <span class="op">=</span> np.array(ds[treatment][line][:i])</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span>:</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>                        ds[treatment][line] <span class="op">=</span> np.concatenate((np.array(ds[treatment][line][:i]), np.array(ds[treatment][line][i<span class="op">+</span><span class="dv">1</span>:])), axis<span class="op">=</span><span class="dv">0</span>)            </span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>                global_i <span class="op">-=</span> <span class="dv">1</span></span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ds</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> remove_cells_two_layer(cells, cell_shapes, lines, treatments, pairwise_dists, ds_proc, ds_align, delete_indices):</span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>    <span class="co">""" </span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a><span class="co">    Remove cells of control group from cells, cell_shapes, ds,</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a><span class="co">    the parameters returned from load_treated_osteosarcoma_cells</span></span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a><span class="co">    Also update n_cells</span></span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a><span class="co">    :param list[int] delete_indices: the indices to delete</span></span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a>    delete_indices <span class="op">=</span> <span class="bu">sorted</span>(delete_indices, reverse<span class="op">=</span><span class="va">True</span>) <span class="co"># to prevent change in index when deleting elements</span></span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Delete elements</span></span>
<span id="cb42-33"><a href="#cb42-33" aria-hidden="true" tabindex="-1"></a>    cells <span class="op">=</span> del_arr_elements(cells, delete_indices)    </span>
<span id="cb42-34"><a href="#cb42-34" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> <span class="bu">list</span>(np.delete(np.array(lines), delete_indices, axis<span class="op">=</span><span class="dv">0</span>))</span>
<span id="cb42-35"><a href="#cb42-35" aria-hidden="true" tabindex="-1"></a>    treatments <span class="op">=</span> <span class="bu">list</span>(np.delete(np.array(treatments), delete_indices, axis<span class="op">=</span><span class="dv">0</span>))</span>
<span id="cb42-36"><a href="#cb42-36" aria-hidden="true" tabindex="-1"></a>    ds_proc <span class="op">=</span> remove_ds_two_layer(ds_proc, delete_indices)</span>
<span id="cb42-37"><a href="#cb42-37" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb42-38"><a href="#cb42-38" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb42-39"><a href="#cb42-39" aria-hidden="true" tabindex="-1"></a>        cell_shapes[metric] <span class="op">=</span> np.delete(np.array(cell_shapes[metric]), delete_indices, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb42-40"><a href="#cb42-40" aria-hidden="true" tabindex="-1"></a>        ds_align[metric] <span class="op">=</span> remove_ds_two_layer(ds_align[metric], delete_indices)</span>
<span id="cb42-41"><a href="#cb42-41" aria-hidden="true" tabindex="-1"></a>        pairwise_dists[metric] <span class="op">=</span> np.delete(pairwise_dists[metric], delete_indices, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb42-42"><a href="#cb42-42" aria-hidden="true" tabindex="-1"></a>        pairwise_dists[metric] <span class="op">=</span> np.delete(pairwise_dists[metric], delete_indices, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb42-43"><a href="#cb42-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-44"><a href="#cb42-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-45"><a href="#cb42-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> cells, cell_shapes, lines, treatments, pairwise_dists, ds_proc, ds_align</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="71230956" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>cells, cell_shapes, lines, treatments, pairwise_dists, ds_proc, ds_align <span class="op">=</span> remove_cells_two_layer(cells, cell_shapes, lines, treatments, pairwise_dists, ds_proc, ds_align, delete_indices)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(85, 2000, 2) (118, 2000, 2)
(18, 2000, 2) (184, 2000, 2)
(86, 1999, 2) (112, 1999, 2)
(19, 1999, 2) (178, 1999, 2)
(86, 1999, 2) (112, 1999, 2)
(19, 1999, 2) (178, 1999, 2)</code></pre>
</div>
</div>
<p>Check we did not loss any other cells after the removal</p>
<div id="abc960d5" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> check_num(cell_shapes, treatments, lines, pairwise_dists, ds_align):</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"treatments number is: </span><span class="sc">{</span><span class="bu">len</span>(treatments)<span class="sc">}</span><span class="ss">, lines number is: </span><span class="sc">{</span><span class="bu">len</span>(lines)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"pairwise_dists for </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> shape is: </span><span class="sc">{</span>pairwise_dists[metric]<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"cell_shapes for </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> number is : </span><span class="sc">{</span><span class="bu">len</span>(cell_shapes[metric])<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"ds_align </span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss"> using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span><span class="bu">len</span>(ds_align[metric][treatment][line])<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ceea0131" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>check_num(cell_shapes, treatments, lines, pairwise_dists, ds_align)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>treatments number is: 623, lines number is: 623
pairwise_dists for SRV shape is: (623, 623)
cell_shapes for SRV number is : 623
ds_align control dlm8 using SRV: 113
ds_align cytd dlm8 using SRV: 74
ds_align jasp dlm8 using SRV: 56
ds_align control dunn using SRV: 197
ds_align cytd dunn using SRV: 92
ds_align jasp dunn using SRV: 91
pairwise_dists for Linear shape is: (623, 623)
cell_shapes for Linear number is : 623
ds_align control dlm8 using Linear: 113
ds_align cytd dlm8 using Linear: 74
ds_align jasp dlm8 using Linear: 56
ds_align control dunn using Linear: 197
ds_align cytd dunn using Linear: 92
ds_align jasp dunn using Linear: 91</code></pre>
</div>
</div>
<p>We compute the mean cell shape by using the SRV metric defined on the space of curves’ shapes. The space of curves’ shape is a manifold: we use the Frechet mean, associated to the SRV metric, to get the mean cell shape.</p>
<p>Do not include cells with duplicate points when calculating the mean shapes</p>
<div id="5933ac2f" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> check_duplicate(cell):</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">""" </span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Return true if there are duplicate points in the cell</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(cell.shape[<span class="dv">0</span>]<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>        cur_coord <span class="op">=</span> cell[i]</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>        next_coord <span class="op">=</span> cell[i<span class="op">+</span><span class="dv">1</span>]</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> np.linalg.norm(cur_coord<span class="op">-</span>next_coord) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="va">True</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Checking the last point vs the first poit</span></span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> np.linalg.norm(cell[<span class="op">-</span><span class="dv">1</span>]<span class="op">-</span>cell[<span class="dv">0</span>]) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">True</span></span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">False</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5110757e" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>delete_indices <span class="op">=</span> []</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, cell <span class="kw">in</span> <span class="bu">reversed</span>(<span class="bu">list</span>(<span class="bu">enumerate</span>(cell_shapes[metric]))):</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> check_duplicate(cell):</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i <span class="kw">not</span> <span class="kw">in</span> delete_indices:</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>                delete_indices.append(i)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>cells, cell_shapes, lines, treatments, pairwise_dists, ds_proc, ds_align <span class="op">=</span> <span class="op">\</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>    remove_cells_two_layer(cells, cell_shapes, lines, treatments, pairwise_dists, ds_proc, ds_align, delete_indices)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Recheck cell number after removing cells with duplicated points</p>
<div id="1ca1d51a" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>check_num(cell_shapes, treatments, lines, pairwise_dists, ds_align)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>treatments number is: 623, lines number is: 623
pairwise_dists for SRV shape is: (623, 623)
cell_shapes for SRV number is : 623
ds_align control dlm8 using SRV: 113
ds_align cytd dlm8 using SRV: 74
ds_align jasp dlm8 using SRV: 56
ds_align control dunn using SRV: 197
ds_align cytd dunn using SRV: 92
ds_align jasp dunn using SRV: 91
pairwise_dists for Linear shape is: (623, 623)
cell_shapes for Linear number is : 623
ds_align control dlm8 using Linear: 113
ds_align cytd dlm8 using Linear: 74
ds_align jasp dlm8 using Linear: 56
ds_align control dunn using Linear: 197
ds_align cytd dunn using Linear: 92
ds_align jasp dunn using Linear: 91</code></pre>
</div>
</div>
<div id="6d2f5abd" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> geomstats.learning.frechet_mean <span class="im">import</span> FrechetMean</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>CURVES_SPACE_SRV <span class="op">=</span> DiscreteCurvesStartingAtOrigin(ambient_dim<span class="op">=</span><span class="dv">2</span>, k_sampling_points<span class="op">=</span>k_sampling_points)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>mean <span class="op">=</span> FrechetMean(CURVES_SPACE_SRV)</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cell_shapes[metric].shape)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>cells <span class="op">=</span> cell_shapes[metric]</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>mean.fit(cells)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>mean_estimate <span class="op">=</span> mean.estimate_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(623, 1999, 2)</code></pre>
</div>
</div>
<div id="b0cbaecb" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>mean_estimate_aligned <span class="op">=</span> {}</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>mean_estimate_clean <span class="op">=</span> mean_estimate[<span class="op">~</span>gs.isnan(gs.<span class="bu">sum</span>(mean_estimate, axis<span class="op">=</span><span class="dv">1</span>)), :]</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>mean_estimate_aligned[metric] <span class="op">=</span> (</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>    mean_estimate_clean <span class="op">-</span> gs.mean(mean_estimate_clean, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Also we compute the linear mean</p>
<div id="72b57862" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'Linear'</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>linear_mean_estimate <span class="op">=</span> gs.mean(cell_shapes[metric], axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>linear_mean_estimate_clean <span class="op">=</span> linear_mean_estimate[<span class="op">~</span>gs.isnan(gs.<span class="bu">sum</span>(linear_mean_estimate, axis<span class="op">=</span><span class="dv">1</span>)), :]</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>mean_estimate_aligned[metric] <span class="op">=</span>  (</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>    linear_mean_estimate_clean <span class="op">-</span> gs.mean(linear_mean_estimate_clean, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot SRV mean cell versus linear mean cell</p>
<div id="28731587" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">6</span>, <span class="dv">3</span>))</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">121</span>)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>plt.plot(mean_estimate_aligned[metric][:, <span class="dv">0</span>], mean_estimate_aligned[metric][:, <span class="dv">1</span>])</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"SRV"</span>)</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"off"</span>)</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>fig.add_subplot(<span class="dv">122</span>)</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'Linear'</span></span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>plt.plot(mean_estimate_aligned[metric][:, <span class="dv">0</span>], mean_estimate_aligned[metric][:, <span class="dv">1</span>])</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Linear"</span>)</span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a>plt.axis(<span class="st">"off"</span>)</span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb56-18"><a href="#cb56-18" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"global_mean.svg"</span>))</span>
<span id="cb56-19"><a href="#cb56-19" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="st">"global_mean.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-43-output-1.png" width="466" height="261" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="analyze-distances-to-the-global-mean-shape" class="level1">
<h1>Analyze Distances to the “Global” Mean Shape</h1>
<p>We consider each of the subgroups of cells, defined by their treatment and cell line. We wish to study how far each of this group is from the global mean shape. We compute the list of distances to the global mean shape.</p>
<div id="ab252440" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>dists_to_global_mean <span class="op">=</span> {}</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>dists_to_global_mean_list <span class="op">=</span> {}</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mean_estimate_aligned[metric].shape)</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>dists_to_global_mean[metric] <span class="op">=</span> apply_func_to_ds(</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>    ds_align[metric], </span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>    func<span class="op">=</span><span class="kw">lambda</span> x: CURVES_SPACE_SRV.metric.dist(x, mean_estimate_aligned[metric])</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>dists_to_global_mean_list[metric] <span class="op">=</span> []</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> t <span class="kw">in</span> TREATMENTS:</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> l <span class="kw">in</span> LINES:</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>        dists_to_global_mean_list[metric].extend(dists_to_global_mean[metric][t][l])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(1999, 2)</code></pre>
</div>
</div>
<p>Compute distances to linear mean</p>
<div id="6d5bc161" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'Linear'</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>dists_to_global_mean[metric] <span class="op">=</span> apply_func_to_ds(</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>    ds_align[metric], func<span class="op">=</span><span class="kw">lambda</span> x: gs.linalg.norm(mean_estimate_aligned[metric] <span class="op">-</span> x) </span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>dists_to_global_mean_list[metric] <span class="op">=</span> []</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> t <span class="kw">in</span> TREATMENTS:</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> l <span class="kw">in</span> LINES:</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>        dists_to_global_mean_list[metric].extend(dists_to_global_mean[metric][t][l])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4a2079bc" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, tight_layout<span class="op">=</span><span class="va">True</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>))</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dlm8'</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>kde_dict <span class="op">=</span> {}</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>    distances <span class="op">=</span> []</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>    min_dists <span class="op">=</span> <span class="bu">min</span>(dists_to_global_mean_list[metric])</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>    max_dists <span class="op">=</span> <span class="bu">max</span>(dists_to_global_mean_list[metric])</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>    xx <span class="op">=</span> gs.linspace(gs.floor(min_dists), gs.ceil(max_dists), k_sampling_points)</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>    kde_dict[metric] <span class="op">=</span> {}</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>        distances <span class="op">=</span> dists_to_global_mean[metric][treatment][line][<span class="op">~</span>gs.isnan(dists_to_global_mean[metric][treatment][line])]</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a>        axs[j].hist(distances, bins<span class="op">=</span><span class="dv">20</span>, alpha<span class="op">=</span><span class="fl">0.4</span>, density<span class="op">=</span><span class="va">True</span>, label<span class="op">=</span>treatment, color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>        kde <span class="op">=</span> stats.gaussian_kde(distances)</span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a>        kde_dict[metric][treatment] <span class="op">=</span> kde</span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a>        axs[j].plot(xx, kde(xx), color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a>        axs[j].set_xlim((min_dists, max_dists))</span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a>        axs[j].legend(fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a>        axs[j].set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a>        axs[j].set_ylabel(<span class="st">"Fraction of cells"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-25"><a href="#cb60-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-26"><a href="#cb60-26" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle("Histograms of SRV distances to global mean cell", fontsize=20)</span></span>
<span id="cb60-27"><a href="#cb60-27" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-28"><a href="#cb60-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb60-29"><a href="#cb60-29" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_histogram.svg"</span>))</span>
<span id="cb60-30"><a href="#cb60-30" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_histogram.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-46-output-1.png" width="758" height="374" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Calculate the ratio of overlapping regions formed by the kde curves</p>
<div id="d6e31cba" class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calc_ratio(kde1, kde2, <span class="bu">min</span>, <span class="bu">max</span>):</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>    xx <span class="op">=</span> np.linspace(<span class="bu">min</span>, <span class="bu">max</span>, <span class="dv">1000</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>    kde1_values <span class="op">=</span> kde1(xx)</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>    kde2_values <span class="op">=</span> kde2(xx)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>    overlap <span class="op">=</span> np.minimum(kde1_values, kde2_values)</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>    overlap_area <span class="op">=</span> np.trapz(overlap, xx)</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>    bound <span class="op">=</span> np.maximum(kde1_values, kde2_values)</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>    bound_area <span class="op">=</span>  np.trapz(bound, xx)</span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> overlap_area<span class="op">/</span>bound_area</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="1a697c4b" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>    min_dists <span class="op">=</span> <span class="bu">min</span>(dists_to_global_mean_list[metric])</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>    max_dists <span class="op">=</span> <span class="bu">max</span>(dists_to_global_mean_list[metric])</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, tmt1 <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(i<span class="op">+</span><span class="dv">1</span>, <span class="bu">len</span>(TREATMENTS)):</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>            tmt2 <span class="op">=</span> TREATMENTS[j]</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>            ratio <span class="op">=</span> calc_ratio(kde_dict[metric][tmt1], kde_dict[metric][tmt2], min_dists, max_dists)</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"Overlap ratio for </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss"> between </span><span class="sc">{</span>tmt1<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>tmt2<span class="sc">}</span><span class="ss"> using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> metric is: </span><span class="sc">{</span><span class="bu">round</span>(ratio, <span class="dv">2</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Overlap ratio for dlm8 between control and cytd using SRV metric is: 0.28
Overlap ratio for dlm8 between control and jasp using SRV metric is: 0.53
Overlap ratio for dlm8 between cytd and jasp using SRV metric is: 0.39
Overlap ratio for dlm8 between control and cytd using Linear metric is: 0.43
Overlap ratio for dlm8 between control and jasp using Linear metric is: 0.69
Overlap ratio for dlm8 between cytd and jasp using Linear metric is: 0.59</code></pre>
</div>
</div>
<div id="f2feee62" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, tight_layout<span class="op">=</span><span class="va">True</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>))</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dunn'</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>np.set_printoptions(precision<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>kde_dict <span class="op">=</span> {}</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>    distances <span class="op">=</span> []</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>    min_dists <span class="op">=</span> <span class="bu">min</span>(dists_to_global_mean_list[metric])</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>    max_dists <span class="op">=</span> <span class="bu">max</span>(dists_to_global_mean_list[metric])</span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>    xx <span class="op">=</span> gs.linspace(gs.floor(min_dists), gs.ceil(max_dists), k_sampling_points)</span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>    kde_dict[metric] <span class="op">=</span> {}</span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a>        distances <span class="op">=</span> dists_to_global_mean[metric][treatment][line][<span class="op">~</span>gs.isnan(dists_to_global_mean[metric][treatment][line])]</span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a>        counts, bin_edges, _ <span class="op">=</span> axs[j].hist(distances, bins<span class="op">=</span><span class="dv">20</span>, alpha<span class="op">=</span><span class="fl">0.4</span>, density<span class="op">=</span><span class="va">True</span>, label<span class="op">=</span>treatment, color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb64-19"><a href="#cb64-19" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(treatment, metric)</span>
<span id="cb64-20"><a href="#cb64-20" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"counts are:"</span>, counts)</span>
<span id="cb64-21"><a href="#cb64-21" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"bin_edges are:"</span>, bin_edges)</span>
<span id="cb64-22"><a href="#cb64-22" aria-hidden="true" tabindex="-1"></a>        kde <span class="op">=</span> stats.gaussian_kde(distances)</span>
<span id="cb64-23"><a href="#cb64-23" aria-hidden="true" tabindex="-1"></a>        kde_dict[metric][treatment] <span class="op">=</span> kde</span>
<span id="cb64-24"><a href="#cb64-24" aria-hidden="true" tabindex="-1"></a>        axs[j].plot(xx, kde(xx), color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb64-25"><a href="#cb64-25" aria-hidden="true" tabindex="-1"></a>        axs[j].set_xlim((min_dists, max_dists))</span>
<span id="cb64-26"><a href="#cb64-26" aria-hidden="true" tabindex="-1"></a>        axs[j].legend(fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb64-27"><a href="#cb64-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-28"><a href="#cb64-28" aria-hidden="true" tabindex="-1"></a>        axs[j].set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb64-29"><a href="#cb64-29" aria-hidden="true" tabindex="-1"></a>        axs[j].set_ylabel(<span class="st">"Fraction of cells"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb64-30"><a href="#cb64-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-31"><a href="#cb64-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-32"><a href="#cb64-32" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle("Histograms of SRV distances to global mean cell", fontsize=20)</span></span>
<span id="cb64-33"><a href="#cb64-33" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb64-34"><a href="#cb64-34" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb64-35"><a href="#cb64-35" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_histogram.svg"</span>))</span>
<span id="cb64-36"><a href="#cb64-36" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_histogram.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>control SRV
counts are: [3.599823688084 9.414923491911 9.138013977443 2.492185630212
 2.215276115744 2.215276115744 2.492185630212 4.15364271702
 6.092009318296 3.876733202552 2.492185630212 1.38454757234
 1.107638057872 2.492185630212 0.553819028936 0.
 0.553819028936 0.             0.             0.276909514468]
bin_edges are: [0.190412844846 0.208744255891 0.227075666936 0.245407077981
 0.263738489026 0.28206990007  0.300401311115 0.31873272216
 0.337064133205 0.35539554425  0.373726955295 0.39205836634
 0.410389777385 0.42872118843  0.447052599475 0.46538401052
 0.483715421565 0.50204683261  0.520378243655 0.5387096547
 0.557041065745]
cytd SRV
counts are: [0.627751614862 0.             1.883254844586 1.255503229724
 1.255503229724 1.255503229724 1.883254844586 5.649764533759
 4.394261304035 5.649764533759 8.160770993208 5.649764533759
 6.905267763483 3.138758074311 2.511006459448 1.255503229724
 3.138758074311 1.883254844586 0.627751614862 0.627751614862]
bin_edges are: [0.26221861859  0.279533691877 0.296848765164 0.314163838451
 0.331478911738 0.348793985025 0.366109058312 0.383424131599
 0.400739204886 0.418054278173 0.43536935146  0.452684424747
 0.469999498034 0.487314571321 0.504629644608 0.521944717895
 0.539259791183 0.55657486447  0.573889937757 0.591205011044
 0.608520084331]
jasp SRV
counts are: [0.928427307436 0.928427307436 0.928427307436 2.785281922307
 3.713709229743 3.713709229743 4.642136537178 6.49899115205
 6.49899115205  6.49899115205  9.284273074357 8.355845766921
 6.49899115205  7.427418459485 2.785281922307 4.642136537178
 1.856854614871 2.785281922307 1.856854614871 1.856854614871]
bin_edges are: [0.244313646946 0.256149803531 0.267985960117 0.279822116702
 0.291658273288 0.303494429873 0.315330586458 0.327166743044
 0.339002899629 0.350839056215 0.3626752128   0.374511369386
 0.386347525971 0.398183682557 0.410019839142 0.421855995727
 0.433692152313 0.445528308898 0.457364465484 0.469200622069
 0.481036778655]
control Linear
counts are: [0.973976940289 1.704459645506 3.895907761156 3.165425055939
 4.626390466373 4.139401996228 5.35687317159  5.843861641734
 4.626390466373 3.408919291012 2.19144811565  2.678436585795
 1.460965410434 0.730482705217 0.973976940289 0.243494235072
 0.730482705217 0.             0.243494235072 0.973976940289]
bin_edges are: [0.084550020208 0.105397093366 0.126244166523 0.147091239681
 0.167938312838 0.188785385996 0.209632459153 0.230479532311
 0.251326605468 0.272173678626 0.293020751783 0.313867824941
 0.334714898098 0.355561971256 0.376409044413 0.397256117571
 0.418103190728 0.438950263886 0.459797337043 0.480644410201
 0.501491483358]
cytd Linear
counts are: [2.686991765509 1.343495882754 1.791327843673 2.239159804591
 2.686991765509 3.582655687345 4.478319609181 4.478319609181
 5.821815491936 4.030487648263 4.030487648263 1.343495882754
 1.343495882754 0.447831960918 0.             0.
 0.447831960918 0.             0.             0.447831960918]
bin_edges are: [0.18370748819  0.20797901449  0.23225054079  0.25652206709
 0.28079359339  0.30506511969  0.32933664599  0.35360817229
 0.37787969859  0.40215122489  0.42642275119  0.45069427749
 0.47496580379  0.49923733009  0.52350885639  0.54778038269
 0.57205190899  0.59632343529  0.62059496159  0.64486648789
 0.669138014189]
jasp Linear
counts are: [3.47808161386  5.21712242079  2.608561210395 6.956163227719
 6.521403025987 6.086642824255 3.912841815592 0.434760201732
 1.73904080693  0.434760201732 0.434760201732 0.
 0.434760201732 0.434760201732 0.             0.
 0.             0.             0.434760201732 0.434760201732]
bin_edges are: [0.154345044651 0.179621072552 0.204897100452 0.230173128353
 0.255449156253 0.280725184154 0.306001212054 0.331277239955
 0.356553267855 0.381829295756 0.407105323656 0.432381351557
 0.457657379457 0.482933407358 0.508209435258 0.533485463159
 0.558761491059 0.58403751896  0.60931354686  0.634589574761
 0.659865602661]</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-49-output-2.png" width="758" height="374" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Calculate the ratio of overlapping regions formed by the three kde curves</p>
<div id="696daf26" class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>    min_dists <span class="op">=</span> <span class="bu">min</span>(dists_to_global_mean_list[metric])</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>    max_dists <span class="op">=</span> <span class="bu">max</span>(dists_to_global_mean_list[metric])</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, tmt1 <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(i<span class="op">+</span><span class="dv">1</span>, <span class="bu">len</span>(TREATMENTS)):</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>            tmt2 <span class="op">=</span> TREATMENTS[j]</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>            ratio <span class="op">=</span> calc_ratio(kde_dict[metric][tmt1], kde_dict[metric][tmt2], min_dists, max_dists)</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"Overlap ratio for </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss"> between </span><span class="sc">{</span>tmt1<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>tmt2<span class="sc">}</span><span class="ss"> using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> metric is: </span><span class="sc">{</span><span class="bu">round</span>(ratio, <span class="dv">2</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Overlap ratio for dunn between control and cytd using SRV metric is: 0.2
Overlap ratio for dunn between control and jasp using SRV metric is: 0.4
Overlap ratio for dunn between cytd and jasp using SRV metric is: 0.35
Overlap ratio for dunn between control and cytd using Linear metric is: 0.32
Overlap ratio for dunn between control and jasp using Linear metric is: 0.72
Overlap ratio for dunn between cytd and jasp using Linear metric is: 0.37</code></pre>
</div>
</div>
<p>Conduct T-test to test if the two samples have the same expected average</p>
<div id="34e76f76" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(TREATMENTS)):</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>        tmt1 <span class="op">=</span> TREATMENTS[i]</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(i<span class="op">+</span><span class="dv">1</span>, <span class="bu">len</span>(TREATMENTS)):</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>            tmt2 <span class="op">=</span> TREATMENTS[j]</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>                distance1 <span class="op">=</span> dists_to_global_mean[metric][tmt1][line][<span class="op">~</span>gs.isnan(dists_to_global_mean[metric][tmt1][line])]</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>                distance2 <span class="op">=</span> dists_to_global_mean[metric][tmt2][line][<span class="op">~</span>gs.isnan(dists_to_global_mean[metric][tmt2][line])]</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>                t_statistic, p_value <span class="op">=</span> stats.ttest_ind(distance1, distance2)</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"Significance of differences for </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss"> between </span><span class="sc">{</span>tmt1<span class="sc">}</span><span class="ss"> and </span><span class="sc">{</span>tmt2<span class="sc">}</span><span class="ss"> using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> metric is: </span><span class="sc">{</span><span class="st">'</span><span class="sc">%.2e</span><span class="st">'</span> <span class="op">%</span> Decimal(p_value)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Significance of differences for dlm8 between control and cytd using SRV metric is: 5.16e-25
Significance of differences for dlm8 between control and cytd using Linear metric is: 3.15e-11
Significance of differences for dlm8 between control and jasp using SRV metric is: 6.87e-06
Significance of differences for dlm8 between control and jasp using Linear metric is: 1.65e-01
Significance of differences for dlm8 between cytd and jasp using SRV metric is: 1.10e-09
Significance of differences for dlm8 between cytd and jasp using Linear metric is: 1.77e-04
Significance of differences for dunn between control and cytd using SRV metric is: 1.29e-41
Significance of differences for dunn between control and cytd using Linear metric is: 3.35e-24
Significance of differences for dunn between control and jasp using SRV metric is: 1.74e-14
Significance of differences for dunn between control and jasp using Linear metric is: 2.50e-03
Significance of differences for dunn between cytd and jasp using SRV metric is: 8.05e-16
Significance of differences for dunn between cytd and jasp using Linear metric is: 1.97e-10</code></pre>
</div>
</div>
<p>Let’s analyze bi-modal distribution for the control group of dunn cell line using SRV metric</p>
<p>We consider two groups: cells with [3.42551653, 3.43015473) - distance to the mean, cells with [3.47189855, 3.47653676) distance to the mean, and find the modes of the two groups</p>
<div id="0a43e5f7" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dunn'</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>treatment <span class="op">=</span> <span class="st">'control'</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>distances <span class="op">=</span> dists_to_global_mean[metric][treatment][line]</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">min</span>(distances), <span class="bu">max</span>(distances))</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>group_1_left <span class="op">=</span> <span class="fl">0.208744255891</span> </span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>group_1_right <span class="op">=</span> <span class="fl">0.227075666936</span></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>group_2_left <span class="op">=</span>  <span class="fl">0.337064133205</span> </span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>group_2_right <span class="op">=</span>  <span class="fl">0.35539554425</span></span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>group_1_indices <span class="op">=</span> [i <span class="cf">for</span> i, element <span class="kw">in</span> <span class="bu">enumerate</span>(distances) <span class="cf">if</span> element <span class="op">&lt;=</span> group_1_right <span class="kw">and</span> element <span class="op">&gt;</span> group_1_left]</span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>group_2_indices <span class="op">=</span> [i <span class="cf">for</span> i, element <span class="kw">in</span> <span class="bu">enumerate</span>(distances) <span class="cf">if</span> element <span class="op">&lt;=</span> group_2_right <span class="kw">and</span> element <span class="op">&gt;</span> group_2_left]</span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(group_1_indices)</span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(group_2_indices)</span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>group_1_cells <span class="op">=</span> gs.array(ds_align[metric][treatment][line])[group_1_indices,:,:]</span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a>group_2_cells <span class="op">=</span> gs.array(ds_align[metric][treatment][line])[group_2_indices,:,:]</span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a>col_num <span class="op">=</span> <span class="bu">max</span>(<span class="bu">len</span>(group_1_indices), <span class="bu">len</span>(group_2_indices))</span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">2</span><span class="op">*</span>col_num, <span class="dv">2</span>))</span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a>count <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> index <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(group_1_indices)):</span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a>    cell <span class="op">=</span> group_1_cells[index]</span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a>    fig.add_subplot(<span class="dv">2</span>, col_num, count)</span>
<span id="cb70-23"><a href="#cb70-23" aria-hidden="true" tabindex="-1"></a>    count <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb70-24"><a href="#cb70-24" aria-hidden="true" tabindex="-1"></a>    plt.plot(cell[:, <span class="dv">0</span>], cell[:, <span class="dv">1</span>])</span>
<span id="cb70-25"><a href="#cb70-25" aria-hidden="true" tabindex="-1"></a>    plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb70-26"><a href="#cb70-26" aria-hidden="true" tabindex="-1"></a>    plt.axis(<span class="st">"off"</span>)</span>
<span id="cb70-27"><a href="#cb70-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-28"><a href="#cb70-28" aria-hidden="true" tabindex="-1"></a>count <span class="op">=</span> <span class="bu">max</span>(<span class="bu">len</span>(group_1_indices), <span class="bu">len</span>(group_2_indices))<span class="op">+</span><span class="dv">1</span></span>
<span id="cb70-29"><a href="#cb70-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> index <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(group_2_indices)):</span>
<span id="cb70-30"><a href="#cb70-30" aria-hidden="true" tabindex="-1"></a>    cell <span class="op">=</span> group_2_cells[index]</span>
<span id="cb70-31"><a href="#cb70-31" aria-hidden="true" tabindex="-1"></a>    fig.add_subplot(<span class="dv">2</span>, col_num, count)</span>
<span id="cb70-32"><a href="#cb70-32" aria-hidden="true" tabindex="-1"></a>    count <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb70-33"><a href="#cb70-33" aria-hidden="true" tabindex="-1"></a>    plt.plot(cell[:, <span class="dv">0</span>], cell[:, <span class="dv">1</span>])</span>
<span id="cb70-34"><a href="#cb70-34" aria-hidden="true" tabindex="-1"></a>    plt.axis(<span class="st">"equal"</span>)</span>
<span id="cb70-35"><a href="#cb70-35" aria-hidden="true" tabindex="-1"></a>    plt.axis(<span class="st">"off"</span>)</span>
<span id="cb70-36"><a href="#cb70-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-37"><a href="#cb70-37" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb70-38"><a href="#cb70-38" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_bimodal_mean.svg"</span>))</span>
<span id="cb70-39"><a href="#cb70-39" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_bimodal_mean.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.19041284484557636 0.5570410657452678
[24, 25, 26, 28, 29, 33, 34, 36, 37, 39, 40, 44, 48, 51, 54, 55, 56, 58, 107, 117, 120, 126, 128, 130, 131, 134, 136, 137, 138, 140, 141, 145, 151, 153]
[2, 10, 16, 17, 64, 66, 77, 80, 86, 87, 90, 91, 95, 100, 101, 104, 157, 167, 168, 170, 175, 196]</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-52-output-2.png" width="5078" height="167" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualization-of-the-mean-of-each-treatment" class="level1">
<h1>Visualization of the Mean of each Treatment</h1>
<p>The mean distances to the global mean shape differ. We also plot the mean shape for each of the subgroup, to get intuition on how the mean shape of each subgroup looks like.</p>
<p>We first calculate the SRV mean</p>
<div id="d09d99ec" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>mean_treatment_cells <span class="op">=</span> {}</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>    treatment_cells <span class="op">=</span> []</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>        treatment_cells.extend(ds_align[metric][treatment][line])</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>    mean_estimator <span class="op">=</span> FrechetMean(space<span class="op">=</span>CURVES_SPACE_SRV)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>    mean_estimator.fit(CURVES_SPACE_SRV.projection(gs.array(treatment_cells)))</span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>    mean_treatment_cells[treatment] <span class="op">=</span> mean_estimator.estimate_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="6a8d625d" class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>mean_line_cells <span class="op">=</span> {}</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>    line_cells <span class="op">=</span> []</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>        line_cells.extend(ds_align[metric][treatment][line])</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>    mean_estimator <span class="op">=</span> FrechetMean(space<span class="op">=</span>CURVES_SPACE_SRV)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>    mean_estimator.fit(CURVES_SPACE_SRV.projection(gs.array(line_cells)))</span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>    mean_line_cells[line] <span class="op">=</span> mean_estimator.estimate_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9284d905" class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>mean_cells <span class="op">=</span> {}</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>mean_cells[metric] <span class="op">=</span> {}</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>    mean_cells[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>        mean_estimator <span class="op">=</span> FrechetMean(space<span class="op">=</span>CURVES_SPACE_SRV)</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>        mean_estimator.fit(CURVES_SPACE_SRV.projection(gs.array(ds_align[metric][treatment][line])))</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>        mean_cells[metric][treatment][line] <span class="op">=</span> mean_estimator.estimate_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then calculate the linear mean</p>
<div id="be1082ba" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'Linear'</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>mean_cells[metric] <span class="op">=</span> {}</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>    mean_cells[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>        mean_cells[metric][treatment][line] <span class="op">=</span> gs.mean(ds_align[metric][treatment][line], axis<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>While the mean shapes of the control groups (for both cell lines) look regular, we observe that: - the mean shape for cytd is the most irregular (for both cell lines) - while the mean shape for jasp is more elongated for dlm8 cell line, and more irregular for dunn cell line.</p>
</section>
<section id="distance-of-the-cell-shapes-to-their-own-mean-shape" class="level1">
<h1>Distance of the Cell Shapes to their Own Mean Shape</h1>
<p>Lastly, we evaluate how each subgroup of cell shapes is distributed around the mean shape of their specific subgroup.</p>
<div id="091389b6" class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>dists_to_own_mean <span class="op">=</span> {}</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>    dists_to_own_mean[metric] <span class="op">=</span> {}</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>        dists_to_own_mean[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>            dists <span class="op">=</span> []</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>            ids <span class="op">=</span> []</span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> i_curve, curve <span class="kw">in</span> <span class="bu">enumerate</span>(ds_align[metric][treatment][line]):</span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> metric <span class="op">==</span> <span class="st">'SRV'</span>:</span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>                    one_dist <span class="op">=</span> CURVES_SPACE_SRV.metric.dist(curve, mean_cells[metric][treatment][line])</span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>                    one_dist <span class="op">=</span> gs.linalg.norm(curve <span class="op">-</span> mean_cells[metric][treatment][line])</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="op">~</span>gs.isnan(one_dist):</span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>                    dists.append(one_dist)</span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a>                    ids.append(i_curve)</span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a>            dists_to_own_mean[metric][treatment][line] <span class="op">=</span> dists</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ddb9dabc" class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Align with ellipse</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dunn'</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="bu">len</span>(TREATMENTS),</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>    nrows<span class="op">=</span><span class="bu">len</span>(METRICS),</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="fl">2.5</span><span class="op">*</span><span class="bu">len</span>(TREATMENTS), <span class="dv">2</span><span class="op">*</span><span class="bu">len</span>(METRICS)))</span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>        ax <span class="op">=</span> axes[j, i]</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>        mean_cell <span class="op">=</span> mean_cells[metric][treatment][line]</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>        ax.plot(mean_cell[:, <span class="dv">0</span>], mean_cell[:, <span class="dv">1</span>], color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>        ax.axis(<span class="st">"equal"</span>)</span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a>        ax.axis(<span class="st">"off"</span>)</span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a>        ax.set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb77-18"><a href="#cb77-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-19"><a href="#cb77-19" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb77-20"><a href="#cb77-20" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_own_mean.svg"</span>))</span>
<span id="cb77-21"><a href="#cb77-21" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_own_mean.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-58-output-1.png" width="587" height="343" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="523b551a" class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dlm8'</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="bu">len</span>(TREATMENTS),</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>    nrows<span class="op">=</span><span class="bu">len</span>(METRICS),</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="fl">2.5</span><span class="op">*</span><span class="bu">len</span>(TREATMENTS), <span class="dv">2</span><span class="op">*</span><span class="bu">len</span>(METRICS)))</span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb78-10"><a href="#cb78-10" aria-hidden="true" tabindex="-1"></a>        ax <span class="op">=</span> axes[j, i]</span>
<span id="cb78-11"><a href="#cb78-11" aria-hidden="true" tabindex="-1"></a>        mean_cell <span class="op">=</span> mean_cells[metric][treatment][line]</span>
<span id="cb78-12"><a href="#cb78-12" aria-hidden="true" tabindex="-1"></a>        ax.plot(mean_cell[:, <span class="dv">0</span>], mean_cell[:, <span class="dv">1</span>], color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb78-13"><a href="#cb78-13" aria-hidden="true" tabindex="-1"></a>        ax.axis(<span class="st">"equal"</span>)</span>
<span id="cb78-14"><a href="#cb78-14" aria-hidden="true" tabindex="-1"></a>        ax.axis(<span class="st">"off"</span>)</span>
<span id="cb78-15"><a href="#cb78-15" aria-hidden="true" tabindex="-1"></a>        ax.set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb78-16"><a href="#cb78-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-17"><a href="#cb78-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb78-18"><a href="#cb78-18" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_own_mean.svg"</span>))</span>
<span id="cb78-19"><a href="#cb78-19" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_own_mean.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-59-output-1.png" width="587" height="343" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We observe for the linear mean, the means go narrower as going right. This is caused by the start points for the cells align exactly on the right with the start point of the reference cell.</p>
<p>We notice this artifactual pattern only happens for the linear means (espectially for the cytd group). Can we argue this is an advantage for SRV (reparameterization + SRV mean)?</p>
<p>The above code find a given number of quantiles within the distance’s histogram, using SRV metric and own mean, and plots the corresponding cell, for each treatment and each cell line.</p>
<div id="a3a7f81b" class="cell" data-execution_count="59">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> ss</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dunn'</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>n_quantiles <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>    nrows<span class="op">=</span><span class="bu">len</span>(TREATMENTS)<span class="op">*</span><span class="bu">len</span>(METRICS),</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span>n_quantiles,</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">20</span>, <span class="dv">2</span> <span class="op">*</span> <span class="bu">len</span>(TREATMENTS) <span class="op">*</span> <span class="bu">len</span>(METRICS)),</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a>ranks <span class="op">=</span> {}</span>
<span id="cb79-13"><a href="#cb79-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-14"><a href="#cb79-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb79-15"><a href="#cb79-15" aria-hidden="true" tabindex="-1"></a>    ranks[treatment] <span class="op">=</span> {}</span>
<span id="cb79-16"><a href="#cb79-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb79-17"><a href="#cb79-17" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb79-18"><a href="#cb79-18" aria-hidden="true" tabindex="-1"></a>        dists_list <span class="op">=</span> dists_to_own_mean[metric][treatment][line]</span>
<span id="cb79-19"><a href="#cb79-19" aria-hidden="true" tabindex="-1"></a>        dists_list <span class="op">=</span> [d <span class="op">+</span> <span class="fl">0.0001</span> <span class="op">*</span> gs.random.rand(<span class="dv">1</span>)[<span class="dv">0</span>] <span class="cf">for</span> d <span class="kw">in</span> dists_list]</span>
<span id="cb79-20"><a href="#cb79-20" aria-hidden="true" tabindex="-1"></a>        cells_list <span class="op">=</span> <span class="bu">list</span>(ds_align[metric][treatment][line])</span>
<span id="cb79-21"><a href="#cb79-21" aria-hidden="true" tabindex="-1"></a>        <span class="cf">assert</span> <span class="bu">len</span>(dists_list) <span class="op">==</span> <span class="bu">len</span>(cells_list)</span>
<span id="cb79-22"><a href="#cb79-22" aria-hidden="true" tabindex="-1"></a>        n_cells <span class="op">=</span> <span class="bu">len</span>(dists_list)</span>
<span id="cb79-23"><a href="#cb79-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-24"><a href="#cb79-24" aria-hidden="true" tabindex="-1"></a>        ranks[treatment][metric] <span class="op">=</span> ss.rankdata(dists_list)</span>
<span id="cb79-25"><a href="#cb79-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-26"><a href="#cb79-26" aria-hidden="true" tabindex="-1"></a>        zipped_lists <span class="op">=</span> <span class="bu">zip</span>(dists_list, cells_list)</span>
<span id="cb79-27"><a href="#cb79-27" aria-hidden="true" tabindex="-1"></a>        sorted_pairs <span class="op">=</span> <span class="bu">sorted</span>(zipped_lists)</span>
<span id="cb79-28"><a href="#cb79-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-29"><a href="#cb79-29" aria-hidden="true" tabindex="-1"></a>        tuples <span class="op">=</span> <span class="bu">zip</span>(<span class="op">*</span>sorted_pairs)</span>
<span id="cb79-30"><a href="#cb79-30" aria-hidden="true" tabindex="-1"></a>        sorted_dists_list, sorted_cells_list <span class="op">=</span> [<span class="bu">list</span>(t) <span class="cf">for</span> t <span class="kw">in</span> tuples]</span>
<span id="cb79-31"><a href="#cb79-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i_quantile <span class="kw">in</span> <span class="bu">range</span>(n_quantiles):</span>
<span id="cb79-32"><a href="#cb79-32" aria-hidden="true" tabindex="-1"></a>            quantile <span class="op">=</span> <span class="bu">int</span>(<span class="fl">0.1</span> <span class="op">*</span> n_cells <span class="op">*</span> i_quantile)</span>
<span id="cb79-33"><a href="#cb79-33" aria-hidden="true" tabindex="-1"></a>            one_cell <span class="op">=</span> sorted_cells_list[quantile]</span>
<span id="cb79-34"><a href="#cb79-34" aria-hidden="true" tabindex="-1"></a>            ax <span class="op">=</span> axes[<span class="dv">2</span><span class="op">*</span>i<span class="op">+</span>j, i_quantile]</span>
<span id="cb79-35"><a href="#cb79-35" aria-hidden="true" tabindex="-1"></a>            ax.plot(one_cell[:, <span class="dv">0</span>], one_cell[:, <span class="dv">1</span>], c<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb79-36"><a href="#cb79-36" aria-hidden="true" tabindex="-1"></a>            ax.set_title(<span class="ss">f"0.</span><span class="sc">{</span>i_quantile<span class="sc">}</span><span class="ss"> quantile"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb79-37"><a href="#cb79-37" aria-hidden="true" tabindex="-1"></a>            <span class="co"># ax.axis("off")</span></span>
<span id="cb79-38"><a href="#cb79-38" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Turn off tick labels</span></span>
<span id="cb79-39"><a href="#cb79-39" aria-hidden="true" tabindex="-1"></a>            ax.set_yticklabels([])</span>
<span id="cb79-40"><a href="#cb79-40" aria-hidden="true" tabindex="-1"></a>            ax.set_xticklabels([])</span>
<span id="cb79-41"><a href="#cb79-41" aria-hidden="true" tabindex="-1"></a>            ax.set_xticks([])</span>
<span id="cb79-42"><a href="#cb79-42" aria-hidden="true" tabindex="-1"></a>            ax.set_yticks([])</span>
<span id="cb79-43"><a href="#cb79-43" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb79-44"><a href="#cb79-44" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb79-45"><a href="#cb79-45" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"bottom"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb79-46"><a href="#cb79-46" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb79-47"><a href="#cb79-47" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i_quantile <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb79-48"><a href="#cb79-48" aria-hidden="true" tabindex="-1"></a>                ax.set_ylabel(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> - </span><span class="ch">\n</span><span class="ss"> </span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss">"</span>, rotation<span class="op">=</span><span class="dv">90</span>, fontsize<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb79-49"><a href="#cb79-49" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb79-50"><a href="#cb79-50" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.suptitle(f"Quantiles for linear metric using own mean", y=-0.01, fontsize=24)</span></span>
<span id="cb79-51"><a href="#cb79-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-52"><a href="#cb79-52" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb79-53"><a href="#cb79-53" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_quantile.svg"</span>))</span>
<span id="cb79-54"><a href="#cb79-54" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_quantile.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-60-output-1.png" width="1909" height="1142" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We do not observe any clear patterns between the rank of the cells with distances using SRV metric and with the linear metric.</p>
<div id="4e55150a" class="cell" data-execution_count="60">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dlm8'</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>n_quantiles <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>    nrows<span class="op">=</span><span class="bu">len</span>(TREATMENTS)<span class="op">*</span><span class="bu">len</span>(METRICS),</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span>n_quantiles,</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">20</span>, <span class="dv">2</span> <span class="op">*</span> <span class="bu">len</span>(TREATMENTS) <span class="op">*</span> <span class="bu">len</span>(METRICS)),</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>        dists_list <span class="op">=</span> dists_to_own_mean[metric][treatment][line]</span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a>        dists_list <span class="op">=</span> [d <span class="op">+</span> <span class="fl">0.0001</span> <span class="op">*</span> gs.random.rand(<span class="dv">1</span>)[<span class="dv">0</span>] <span class="cf">for</span> d <span class="kw">in</span> dists_list]</span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>        cells_list <span class="op">=</span> <span class="bu">list</span>(ds_align[metric][treatment][line])</span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">assert</span> <span class="bu">len</span>(dists_list) <span class="op">==</span> <span class="bu">len</span>(dists_list)</span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a>        n_cells <span class="op">=</span> <span class="bu">len</span>(dists_list)</span>
<span id="cb80-17"><a href="#cb80-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-18"><a href="#cb80-18" aria-hidden="true" tabindex="-1"></a>        zipped_lists <span class="op">=</span> <span class="bu">zip</span>(dists_list, cells_list)</span>
<span id="cb80-19"><a href="#cb80-19" aria-hidden="true" tabindex="-1"></a>        sorted_pairs <span class="op">=</span> <span class="bu">sorted</span>(zipped_lists)</span>
<span id="cb80-20"><a href="#cb80-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-21"><a href="#cb80-21" aria-hidden="true" tabindex="-1"></a>        tuples <span class="op">=</span> <span class="bu">zip</span>(<span class="op">*</span>sorted_pairs)</span>
<span id="cb80-22"><a href="#cb80-22" aria-hidden="true" tabindex="-1"></a>        sorted_dists_list, sorted_cells_list <span class="op">=</span> [<span class="bu">list</span>(t) <span class="cf">for</span> t <span class="kw">in</span> tuples]</span>
<span id="cb80-23"><a href="#cb80-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i_quantile <span class="kw">in</span> <span class="bu">range</span>(n_quantiles):</span>
<span id="cb80-24"><a href="#cb80-24" aria-hidden="true" tabindex="-1"></a>            quantile <span class="op">=</span> <span class="bu">int</span>(<span class="fl">0.1</span> <span class="op">*</span> n_cells <span class="op">*</span> i_quantile)</span>
<span id="cb80-25"><a href="#cb80-25" aria-hidden="true" tabindex="-1"></a>            one_cell <span class="op">=</span> sorted_cells_list[quantile]</span>
<span id="cb80-26"><a href="#cb80-26" aria-hidden="true" tabindex="-1"></a>            ax <span class="op">=</span> axes[<span class="dv">2</span><span class="op">*</span>i<span class="op">+</span>j, i_quantile]</span>
<span id="cb80-27"><a href="#cb80-27" aria-hidden="true" tabindex="-1"></a>            ax.plot(one_cell[:, <span class="dv">0</span>], one_cell[:, <span class="dv">1</span>], c<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb80-28"><a href="#cb80-28" aria-hidden="true" tabindex="-1"></a>            ax.set_title(<span class="ss">f"0.</span><span class="sc">{</span>i_quantile<span class="sc">}</span><span class="ss"> quantile"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb80-29"><a href="#cb80-29" aria-hidden="true" tabindex="-1"></a>            <span class="co"># ax.axis("off")</span></span>
<span id="cb80-30"><a href="#cb80-30" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Turn off tick labels</span></span>
<span id="cb80-31"><a href="#cb80-31" aria-hidden="true" tabindex="-1"></a>            ax.set_yticklabels([])</span>
<span id="cb80-32"><a href="#cb80-32" aria-hidden="true" tabindex="-1"></a>            ax.set_xticklabels([])</span>
<span id="cb80-33"><a href="#cb80-33" aria-hidden="true" tabindex="-1"></a>            ax.set_xticks([])</span>
<span id="cb80-34"><a href="#cb80-34" aria-hidden="true" tabindex="-1"></a>            ax.set_yticks([])</span>
<span id="cb80-35"><a href="#cb80-35" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb80-36"><a href="#cb80-36" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb80-37"><a href="#cb80-37" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"bottom"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb80-38"><a href="#cb80-38" aria-hidden="true" tabindex="-1"></a>            ax.spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb80-39"><a href="#cb80-39" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i_quantile <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb80-40"><a href="#cb80-40" aria-hidden="true" tabindex="-1"></a>                ax.set_ylabel(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> - </span><span class="ch">\n</span><span class="ss"> </span><span class="sc">{</span>treatment<span class="sc">}</span><span class="ss">"</span>, rotation<span class="op">=</span><span class="dv">90</span>, fontsize<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb80-41"><a href="#cb80-41" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb80-42"><a href="#cb80-42" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.suptitle(f"Quantiles for linear metric using own mean", y=-0.01, fontsize=24)</span></span>
<span id="cb80-43"><a href="#cb80-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-44"><a href="#cb80-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb80-45"><a href="#cb80-45" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_quantile.svg"</span>))</span>
<span id="cb80-46"><a href="#cb80-46" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_quantile.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-61-output-1.png" width="1909" height="1142" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The above code find a given number of quantiles within the distance’s histogram, using linear metric and own mean, and plots the corresponding cell, for each treatment and each cell line.</p>
</section>
<section id="dimensionality-reduction" class="level1">
<h1>Dimensionality Reduction</h1>
<p>We use the following experiments to illustrate how SRV metric can help with dimensionality reduction</p>
<div id="4d5e81de" class="cell" data-execution_count="61">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> scaled_stress(pos, pairwise_dists):</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">""" </span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Calculate the scaled stress invariant to scaling using the original stress </span><span class="ch">\</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co">    statistics and actual pairwise distances</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co">    :param float unscaled_stress: the original stress</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="co">    :param 2D np.array[float] pairwise_dists: pairwise distance</span></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># compute pairwise distance of pos</span></span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>    pairwise_pos <span class="op">=</span> np.empty(shape<span class="op">=</span>(pos.shape[<span class="dv">0</span>], pos.shape[<span class="dv">0</span>]))</span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(pos.shape[<span class="dv">0</span>]):</span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(pos.shape[<span class="dv">0</span>]):</span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a>            pairwise_pos[i,j] <span class="op">=</span> np.sqrt(np.<span class="bu">sum</span>(pos[i]<span class="op">-</span>pos[j])<span class="op">**</span><span class="dv">2</span>)</span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(pairwise_pos)</span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a>    stress <span class="op">=</span> np.sqrt(np.<span class="bu">sum</span>((pairwise_dists<span class="op">-</span>pairwise_pos)<span class="op">**</span><span class="dv">2</span>))</span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> stress<span class="op">/</span>np.sqrt(np.<span class="bu">sum</span>(pairwise_dists<span class="op">**</span><span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f31e131a" class="cell" data-execution_count="62">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> {}</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>pos <span class="op">=</span> {}</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>dims <span class="op">=</span> <span class="bu">range</span>(<span class="dv">2</span>, <span class="dv">11</span>)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>stresses <span class="op">=</span> {}</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>    mds[metric] <span class="op">=</span> {}</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>    pos[metric] <span class="op">=</span> {}</span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>    stresses[metric] <span class="op">=</span> []</span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> dim <span class="kw">in</span> dims:</span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>        mds[metric][dim] <span class="op">=</span> manifold.MDS(n_components<span class="op">=</span>dim, random_state<span class="op">=</span><span class="dv">0</span>, dissimilarity<span class="op">=</span><span class="st">"precomputed"</span>) <span class="co"># random_state set to 10</span></span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>        pos[metric][dim] <span class="op">=</span> mds[metric][dim].fit(pairwise_dists[metric]).embedding_</span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>        stress_val <span class="op">=</span> mds[metric][dim].stress_</span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>        scaled_stress_val <span class="op">=</span> np.sqrt(stress_val<span class="op">/</span>((pairwise_dists[metric]<span class="op">**</span><span class="dv">2</span>).<span class="bu">sum</span>()<span class="op">/</span><span class="dv">2</span>))</span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>        <span class="co"># scaled_stress_val = scaled_stress(pos[metric][dim], pairwise_dists[metric])</span></span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-17"><a href="#cb82-17" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"the unscaled stress for </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> model is for </span><span class="sc">{</span>dim<span class="sc">}</span><span class="ss">:"</span>, stress_val)</span>
<span id="cb82-18"><a href="#cb82-18" aria-hidden="true" tabindex="-1"></a>        stresses[metric].append(scaled_stress_val)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>the unscaled stress for SRV model is for 2: 0.0015505150986308987
the unscaled stress for SRV model is for 3: 0.0009766856050873998
the unscaled stress for SRV model is for 4: 0.0007390199671520337
the unscaled stress for SRV model is for 5: 0.0005748305174444293
the unscaled stress for SRV model is for 6: 0.00047113942181298865
the unscaled stress for SRV model is for 7: 0.0003990770585748401
the unscaled stress for SRV model is for 8: 0.00034641999727906943
the unscaled stress for SRV model is for 9: 0.00030596906074277627
the unscaled stress for SRV model is for 10: 0.00027546016788315334
the unscaled stress for Linear model is for 2: 0.0012568732933103922
the unscaled stress for Linear model is for 3: 0.0008789553123291832
the unscaled stress for Linear model is for 4: 0.0007370740946128706
the unscaled stress for Linear model is for 5: 0.0006365408960217103
the unscaled stress for Linear model is for 6: 0.0005664042865819429
the unscaled stress for Linear model is for 7: 0.0005223292015115522
the unscaled stress for Linear model is for 8: 0.0004846528585517728
the unscaled stress for Linear model is for 9: 0.00046151351278745815
the unscaled stress for Linear model is for 10: 0.0004397214282582284</code></pre>
</div>
</div>
<div id="5cf2dce4" class="cell" data-execution_count="63">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize <span class="op">=</span> (<span class="dv">4</span>,<span class="dv">4</span>))</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>    plt.scatter(dims, stresses[metric], label<span class="op">=</span>metric)</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>    plt.plot(dims, stresses[metric])</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>plt.xticks(dims)</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>plt.legend()</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"MDS_stress.svg"</span>))</span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"MDS_stress.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-64-output-1.png" width="356" height="337" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>In terms of the scaled stress statistics, we observe linear metric perform better than SRV metric. That is, linear metric preserves the pairwise distances in embedded dimension better than the SRV metric.</p>
<p>Calculate MDS statistics for dimension 2</p>
<div id="e6a355c6" class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> manifold.MDS(n_components<span class="op">=</span><span class="dv">2</span>, random_state<span class="op">=</span><span class="dv">0</span>, dissimilarity<span class="op">=</span><span class="st">"precomputed"</span>)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>pos <span class="op">=</span> mds.fit(pairwise_dists[metric]).embedding_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>MDS embedding of cell treatments (control, cytd and jasp) for different cell lines (dunn and dlm8)</p>
<div id="421a858b" class="cell" data-execution_count="65">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>embs <span class="op">=</span> {}</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>embs[metric] <span class="op">=</span> {}</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>    embs[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>        cell_num <span class="op">=</span> <span class="bu">len</span>(ds_align[metric][treatment][line])   </span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>        embs[metric][treatment][line] <span class="op">=</span> pos[index:index<span class="op">+</span>cell_num]</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>        index <span class="op">+=</span> cell_num</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We draw a comparison with linear metric using the following code</p>
<div id="ecab940f" class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'Linear'</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> manifold.MDS(n_components<span class="op">=</span><span class="dv">2</span>, random_state<span class="op">=</span><span class="dv">0</span>, dissimilarity<span class="op">=</span><span class="st">"precomputed"</span>)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>pos <span class="op">=</span> mds.fit(pairwise_dists[metric]).embedding_</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"the stress for linear model is:"</span>, mds.stress_)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>the stress for linear model is: 0.0012568732933103922</code></pre>
</div>
</div>
<div id="f67ac802" class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>embs[metric] <span class="op">=</span> {}</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>    embs[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>        cell_num <span class="op">=</span> <span class="bu">len</span>(ds_align[metric][treatment][line])   </span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>        embs[metric][treatment][line] <span class="op">=</span> pos[index:index<span class="op">+</span>cell_num]</span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>        index <span class="op">+=</span> cell_num</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The stress for MDS embedding using the linear metric is better than SRV metric.</p>
<p>However, if we can make a better interpretation of the visual result of SRV metric, we could still argue SRV is better at capturing cell heterogeneity.</p>
<div id="8a039e06" class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>embs[metric] <span class="op">=</span> {}</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>    embs[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>        cell_num <span class="op">=</span> <span class="bu">len</span>(ds_align[metric][treatment][line])   </span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>        embs[metric][treatment][line] <span class="op">=</span> pos[index:index<span class="op">+</span>cell_num]</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>        index <span class="op">+=</span> cell_num</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="797b77de" class="cell" data-execution_count="69">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, tight_layout<span class="op">=</span><span class="va">True</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>))</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dunn'</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>        cur_embs <span class="op">=</span> embs[metric][treatment][line]</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>        axs[j].scatter(</span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">0</span>],</span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">1</span>],</span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a>                label<span class="op">=</span>treatment,</span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a>                s<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb91-12"><a href="#cb91-12" aria-hidden="true" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.4</span></span>
<span id="cb91-13"><a href="#cb91-13" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb91-14"><a href="#cb91-14" aria-hidden="true" tabindex="-1"></a>        <span class="co"># axs[j].set_xlim(-3.5*1e-5, 3.5*1e-5)</span></span>
<span id="cb91-15"><a href="#cb91-15" aria-hidden="true" tabindex="-1"></a>        axs[j].set_xlabel(<span class="st">"First Dimension"</span>)</span>
<span id="cb91-16"><a href="#cb91-16" aria-hidden="true" tabindex="-1"></a>        axs[j].set_ylabel(<span class="st">"Second Dimension"</span>)</span>
<span id="cb91-17"><a href="#cb91-17" aria-hidden="true" tabindex="-1"></a>        axs[j].legend()</span>
<span id="cb91-18"><a href="#cb91-18" aria-hidden="true" tabindex="-1"></a>        axs[j].set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb91-19"><a href="#cb91-19" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle("MDS of cell shapes using SRV metric", fontsize=20)</span></span>
<span id="cb91-20"><a href="#cb91-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-21"><a href="#cb91-21" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb91-22"><a href="#cb91-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-23"><a href="#cb91-23" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb91-24"><a href="#cb91-24" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_2D.svg"</span>))</span>
<span id="cb91-25"><a href="#cb91-25" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_2D.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-70-output-1.png" width="758" height="374" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="8d430c1c" class="cell" data-execution_count="70">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, tight_layout<span class="op">=</span><span class="va">True</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>))</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dlm8'</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>    distances <span class="op">=</span> []</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>        cur_embs <span class="op">=</span> embs[metric][treatment][line]</span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a>        axs[j].scatter(</span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">0</span>],</span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">1</span>],</span>
<span id="cb92-11"><a href="#cb92-11" aria-hidden="true" tabindex="-1"></a>                label<span class="op">=</span>treatment,</span>
<span id="cb92-12"><a href="#cb92-12" aria-hidden="true" tabindex="-1"></a>                s<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb92-13"><a href="#cb92-13" aria-hidden="true" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.4</span></span>
<span id="cb92-14"><a href="#cb92-14" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb92-15"><a href="#cb92-15" aria-hidden="true" tabindex="-1"></a>        <span class="co"># axs[j].set_xlim(-3.5*1e-5, 3.5*1e-5)</span></span>
<span id="cb92-16"><a href="#cb92-16" aria-hidden="true" tabindex="-1"></a>        axs[j].set_xlabel(<span class="st">"First Dimension"</span>)</span>
<span id="cb92-17"><a href="#cb92-17" aria-hidden="true" tabindex="-1"></a>        axs[j].set_ylabel(<span class="st">"Second Dimension"</span>)</span>
<span id="cb92-18"><a href="#cb92-18" aria-hidden="true" tabindex="-1"></a>        axs[j].legend()</span>
<span id="cb92-19"><a href="#cb92-19" aria-hidden="true" tabindex="-1"></a>        axs[j].set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb92-20"><a href="#cb92-20" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle("MDS of cell shapes using SRV metric", fontsize=20)</span></span>
<span id="cb92-21"><a href="#cb92-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-22"><a href="#cb92-22" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb92-23"><a href="#cb92-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-24"><a href="#cb92-24" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb92-25"><a href="#cb92-25" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_2D.svg"</span>))</span>
<span id="cb92-26"><a href="#cb92-26" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_2D.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-71-output-1.png" width="758" height="374" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We also consider embedding in 3D.</p>
<div id="f88e2012" class="cell" data-execution_count="71">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'SRV'</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> manifold.MDS(n_components<span class="op">=</span><span class="dv">3</span>, random_state<span class="op">=</span><span class="dv">0</span>, dissimilarity<span class="op">=</span><span class="st">"precomputed"</span>)</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>pos <span class="op">=</span> mds.fit(pairwise_dists[metric]).embedding_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="1decacef" class="cell" data-execution_count="72">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>embs <span class="op">=</span> {}</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>embs[metric] <span class="op">=</span> {}</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>    embs[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>        cell_num <span class="op">=</span> <span class="bu">len</span>(ds_align[metric][treatment][line])   </span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>        embs[metric][treatment][line] <span class="op">=</span> pos[index:index<span class="op">+</span>cell_num]</span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>        index <span class="op">+=</span> cell_num</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="60ebfbcb" class="cell" data-execution_count="73">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>metric <span class="op">=</span> <span class="st">'Linear'</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>mds <span class="op">=</span> manifold.MDS(n_components<span class="op">=</span><span class="dv">3</span>, random_state<span class="op">=</span><span class="dv">1</span>, dissimilarity<span class="op">=</span><span class="st">"precomputed"</span>)</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>pos <span class="op">=</span> mds.fit(pairwise_dists[metric]).embedding_</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"the stress for linear model is:"</span>, mds.stress_)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>the stress for linear model is: 0.0008821306413255005</code></pre>
</div>
</div>
<div id="a41cd3ca" class="cell" data-execution_count="74">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>embs[metric] <span class="op">=</span> {}</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>index <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treatment <span class="kw">in</span> TREATMENTS:</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>    embs[metric][treatment] <span class="op">=</span> {}</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>        cell_num <span class="op">=</span> <span class="bu">len</span>(ds_align[metric][treatment][line])   </span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>        embs[metric][treatment][line] <span class="op">=</span> pos[index:index<span class="op">+</span>cell_num]</span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>        index <span class="op">+=</span> cell_num</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="57329585" class="cell" data-execution_count="75">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, tight_layout<span class="op">=</span><span class="va">True</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>), subplot_kw<span class="op">=</span><span class="bu">dict</span>(projection<span class="op">=</span><span class="st">'3d'</span>))</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dunn'</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>    distances <span class="op">=</span> []</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>        cur_embs <span class="op">=</span> embs[metric][treatment][line]</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>        axs[j].scatter(</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">0</span>],</span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">1</span>],</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">2</span>],</span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a>                label<span class="op">=</span>treatment,</span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>                s<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb98-14"><a href="#cb98-14" aria-hidden="true" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.4</span></span>
<span id="cb98-15"><a href="#cb98-15" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb98-16"><a href="#cb98-16" aria-hidden="true" tabindex="-1"></a>        axs[j].set_xlabel(<span class="st">"First Dimension"</span>)</span>
<span id="cb98-17"><a href="#cb98-17" aria-hidden="true" tabindex="-1"></a>        axs[j].set_ylabel(<span class="st">"Second Dimension"</span>)</span>
<span id="cb98-18"><a href="#cb98-18" aria-hidden="true" tabindex="-1"></a>        axs[j].legend()</span>
<span id="cb98-19"><a href="#cb98-19" aria-hidden="true" tabindex="-1"></a>        axs[j].set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb98-20"><a href="#cb98-20" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle("MDS of cell shapes using linear metric", fontsize=20)</span></span>
<span id="cb98-21"><a href="#cb98-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-22"><a href="#cb98-22" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb98-23"><a href="#cb98-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-24"><a href="#cb98-24" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb98-25"><a href="#cb98-25" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_3D.svg"</span>))</span>
<span id="cb98-26"><a href="#cb98-26" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_3D.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-76-output-1.png" width="777" height="398" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="924ba931" class="cell" data-execution_count="76">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, tight_layout<span class="op">=</span><span class="va">True</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">4</span>), subplot_kw<span class="op">=</span><span class="bu">dict</span>(projection<span class="op">=</span><span class="st">'3d'</span>))</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>line <span class="op">=</span> <span class="st">'dlm8'</span></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> j, metric <span class="kw">in</span> <span class="bu">enumerate</span>(METRICS):</span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>    distances <span class="op">=</span> []</span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, treatment <span class="kw">in</span> <span class="bu">enumerate</span>(TREATMENTS):</span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a>        cur_embs <span class="op">=</span> embs[metric][treatment][line]</span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a>        axs[j].scatter(</span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">0</span>],</span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">1</span>],</span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a>                cur_embs[:, <span class="dv">2</span>],</span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a>                label<span class="op">=</span>treatment,</span>
<span id="cb99-13"><a href="#cb99-13" aria-hidden="true" tabindex="-1"></a>                s<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb99-14"><a href="#cb99-14" aria-hidden="true" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.4</span></span>
<span id="cb99-15"><a href="#cb99-15" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb99-16"><a href="#cb99-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># axs[j].set_xlim(-3.5*1e-5, 3.5*1e-5)</span></span>
<span id="cb99-17"><a href="#cb99-17" aria-hidden="true" tabindex="-1"></a>        axs[j].set_xlabel(<span class="st">"First Dimension"</span>)</span>
<span id="cb99-18"><a href="#cb99-18" aria-hidden="true" tabindex="-1"></a>        axs[j].set_ylabel(<span class="st">"Second Dimension"</span>)</span>
<span id="cb99-19"><a href="#cb99-19" aria-hidden="true" tabindex="-1"></a>        axs[j].legend()</span>
<span id="cb99-20"><a href="#cb99-20" aria-hidden="true" tabindex="-1"></a>        axs[j].set_title(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb99-21"><a href="#cb99-21" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle("MDS of cell shapes using linear metric", fontsize=20)</span></span>
<span id="cb99-22"><a href="#cb99-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-23"><a href="#cb99-23" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb99-24"><a href="#cb99-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-25"><a href="#cb99-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> savefig:</span>
<span id="cb99-26"><a href="#cb99-26" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_3D.svg"</span>))</span>
<span id="cb99-27"><a href="#cb99-27" aria-hidden="true" tabindex="-1"></a>    plt.savefig(os.path.join(figs_dir, <span class="ss">f"</span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">_MDS_3D.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="osteosarcoma_analysis_files/figure-html/cell-77-output-1.png" width="777" height="394" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="multi-class-3-class-classification" class="level1">
<h1>Multi-class (3-class) classification</h1>
<p>We now consider one cell line at the same time, to investigate the effects of the drugs on the cell shapes. Applying the MDS again gives the following results:</p>
<p>Since the detected subspace dimension for this dataset is 3, we perform the classification based on 3D embeddings.</p>
<div id="4772b0f3" class="cell" data-execution_count="77">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> precision_score, recall_score, accuracy_score</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> svm_5_fold_classification(X, y):</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Initialize a Support Vector Classifier</span></span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>    svm_classifier <span class="op">=</span> svm.SVC(kernel<span class="op">=</span><span class="st">'poly'</span>, degree<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Prepare to split the data into 5 folds, maintaining the percentage of samples for each class</span></span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>    skf <span class="op">=</span> StratifiedKFold(n_splits<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># To store precision and recall per class for each fold</span></span>
<span id="cb100-11"><a href="#cb100-11" aria-hidden="true" tabindex="-1"></a>    precisions_per_class <span class="op">=</span> []</span>
<span id="cb100-12"><a href="#cb100-12" aria-hidden="true" tabindex="-1"></a>    recalls_per_class <span class="op">=</span> []</span>
<span id="cb100-13"><a href="#cb100-13" aria-hidden="true" tabindex="-1"></a>    accuracy_per_class <span class="op">=</span> []</span>
<span id="cb100-14"><a href="#cb100-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-15"><a href="#cb100-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Perform 5-fold cross-validation</span></span>
<span id="cb100-16"><a href="#cb100-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> train_index, test_index <span class="kw">in</span> skf.split(X, y):</span>
<span id="cb100-17"><a href="#cb100-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Splitting data into training and test sets</span></span>
<span id="cb100-18"><a href="#cb100-18" aria-hidden="true" tabindex="-1"></a>        X_train, X_test <span class="op">=</span> X[train_index], X[test_index]</span>
<span id="cb100-19"><a href="#cb100-19" aria-hidden="true" tabindex="-1"></a>        y_train, y_test <span class="op">=</span> y[train_index], y[test_index]</span>
<span id="cb100-20"><a href="#cb100-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-21"><a href="#cb100-21" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Train the model</span></span>
<span id="cb100-22"><a href="#cb100-22" aria-hidden="true" tabindex="-1"></a>        svm_classifier.fit(X_train, y_train)</span>
<span id="cb100-23"><a href="#cb100-23" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb100-24"><a href="#cb100-24" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Predict on the test data</span></span>
<span id="cb100-25"><a href="#cb100-25" aria-hidden="true" tabindex="-1"></a>        y_pred <span class="op">=</span> svm_classifier.predict(X_test)</span>
<span id="cb100-26"><a href="#cb100-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-27"><a href="#cb100-27" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate precision and recall per class</span></span>
<span id="cb100-28"><a href="#cb100-28" aria-hidden="true" tabindex="-1"></a>        precision <span class="op">=</span> precision_score(y_test, y_pred, average<span class="op">=</span><span class="va">None</span>, zero_division<span class="op">=</span>np.nan)</span>
<span id="cb100-29"><a href="#cb100-29" aria-hidden="true" tabindex="-1"></a>        recall <span class="op">=</span> recall_score(y_test, y_pred, average<span class="op">=</span><span class="va">None</span>, zero_division<span class="op">=</span>np.nan)</span>
<span id="cb100-30"><a href="#cb100-30" aria-hidden="true" tabindex="-1"></a>        accuracy <span class="op">=</span> accuracy_score(y_test, y_pred)</span>
<span id="cb100-31"><a href="#cb100-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-32"><a href="#cb100-32" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Store results from each fold</span></span>
<span id="cb100-33"><a href="#cb100-33" aria-hidden="true" tabindex="-1"></a>        precisions_per_class.append(precision)</span>
<span id="cb100-34"><a href="#cb100-34" aria-hidden="true" tabindex="-1"></a>        recalls_per_class.append(recall)</span>
<span id="cb100-35"><a href="#cb100-35" aria-hidden="true" tabindex="-1"></a>        accuracy_per_class.append(accuracy)</span>
<span id="cb100-36"><a href="#cb100-36" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb100-37"><a href="#cb100-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate the mean precision and recall per class across all folds</span></span>
<span id="cb100-38"><a href="#cb100-38" aria-hidden="true" tabindex="-1"></a>    mean_precisions <span class="op">=</span> np.mean(precisions_per_class, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb100-39"><a href="#cb100-39" aria-hidden="true" tabindex="-1"></a>    mean_recalls <span class="op">=</span> np.mean(recalls_per_class, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb100-40"><a href="#cb100-40" aria-hidden="true" tabindex="-1"></a>    mean_accuracies <span class="op">=</span> np.mean(accuracy_per_class, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb100-41"><a href="#cb100-41" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb100-42"><a href="#cb100-42" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Mean precisions per class across all folds:"</span>, <span class="bu">round</span>(np.mean(mean_precisions), <span class="dv">2</span>))</span>
<span id="cb100-43"><a href="#cb100-43" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Mean recalls per class across all folds:"</span>, <span class="bu">round</span>(np.mean(mean_recalls), <span class="dv">2</span>))</span>
<span id="cb100-44"><a href="#cb100-44" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Mean accuracies per class across all folds:"</span>, <span class="bu">round</span>(mean_accuracies, <span class="dv">2</span>))</span>
<span id="cb100-45"><a href="#cb100-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-46"><a href="#cb100-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> mean_precisions, mean_recalls</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5321344d" class="cell" data-execution_count="78">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>lines <span class="op">=</span> gs.array(lines)</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>treatments <span class="op">=</span> gs.array(treatments)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d990e91e" class="cell" data-execution_count="79">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> LINES:</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> metric <span class="kw">in</span> METRICS:</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>        control_indexes <span class="op">=</span> gs.where((lines <span class="op">==</span> line) <span class="op">&amp;</span> (treatments <span class="op">==</span> <span class="st">"control"</span>))[<span class="dv">0</span>]</span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>        cytd_indexes <span class="op">=</span> gs.where((lines <span class="op">==</span> line) <span class="op">&amp;</span> (treatments <span class="op">==</span> <span class="st">"cytd"</span>))[<span class="dv">0</span>]</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>        jasp_indexes <span class="op">=</span> gs.where((lines <span class="op">==</span> line) <span class="op">&amp;</span> (treatments <span class="op">==</span> <span class="st">"jasp"</span>))[<span class="dv">0</span>]</span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>        treatment_indexes <span class="op">=</span> gs.where((lines <span class="op">==</span> line) <span class="op">&amp;</span> (treatments <span class="op">!=</span> <span class="st">'control'</span>))[<span class="dv">0</span>]</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>        <span class="co"># indexes = gs.concatenate((jasp_indexes, cytd_indexes, control_indexes))</span></span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>        indexes <span class="op">=</span> gs.concatenate((control_indexes, treatment_indexes))</span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a>        matrix <span class="op">=</span> pairwise_dists[metric][indexes][:, indexes]</span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a>        mds <span class="op">=</span> manifold.MDS(n_components<span class="op">=</span><span class="dv">2</span>, random_state <span class="op">=</span> <span class="dv">10</span>, dissimilarity<span class="op">=</span><span class="st">"precomputed"</span>)</span>
<span id="cb102-13"><a href="#cb102-13" aria-hidden="true" tabindex="-1"></a>        pos <span class="op">=</span> mds.fit(matrix).embedding_</span>
<span id="cb102-14"><a href="#cb102-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-15"><a href="#cb102-15" aria-hidden="true" tabindex="-1"></a>        line_treatments <span class="op">=</span> treatments[lines <span class="op">==</span> line]</span>
<span id="cb102-16"><a href="#cb102-16" aria-hidden="true" tabindex="-1"></a>        line_treatments_strings, line_treatments_labels <span class="op">=</span> np.unique(line_treatments, return_inverse<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb102-17"><a href="#cb102-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># print(line_treatments_strings)</span></span>
<span id="cb102-18"><a href="#cb102-18" aria-hidden="true" tabindex="-1"></a>        <span class="co"># print(line_treatments_labels)</span></span>
<span id="cb102-19"><a href="#cb102-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-20"><a href="#cb102-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i, label <span class="kw">in</span> <span class="bu">enumerate</span>(line_treatments_labels):</span>
<span id="cb102-21"><a href="#cb102-21" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> line_treatments_strings[label] <span class="op">==</span> <span class="st">'cytd'</span> <span class="kw">or</span> line_treatments_strings[label] <span class="op">==</span> <span class="st">'jasp'</span>:</span>
<span id="cb102-22"><a href="#cb102-22" aria-hidden="true" tabindex="-1"></a>                line_treatments_labels[i] <span class="op">=</span> <span class="bu">len</span>(line_treatments_strings)</span>
<span id="cb102-23"><a href="#cb102-23" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb102-24"><a href="#cb102-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-25"><a href="#cb102-25" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Using </span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss"> on </span><span class="sc">{</span>line<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb102-26"><a href="#cb102-26" aria-hidden="true" tabindex="-1"></a>        <span class="co"># print(line_treatments_labels)</span></span>
<span id="cb102-27"><a href="#cb102-27" aria-hidden="true" tabindex="-1"></a>        svm_5_fold_classification(pos, line_treatments_labels)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Using SRV on dlm8
Mean precisions per class across all folds: 0.71
Mean recalls per class across all folds: 0.7
Mean accuracies per class across all folds: 0.69
Using Linear on dlm8
Mean precisions per class across all folds: 0.68
Mean recalls per class across all folds: 0.62
Mean accuracies per class across all folds: 0.6
Using SRV on dunn
Mean precisions per class across all folds: 0.73
Mean recalls per class across all folds: 0.69
Mean accuracies per class across all folds: 0.7
Using Linear on dunn
Mean precisions per class across all folds: 0.62
Mean recalls per class across all folds: 0.59
Mean accuracies per class across all folds: 0.6</code></pre>
</div>
</div>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-li2023using" class="csl-entry" role="listitem">
Li, Wanxin, Ashok Prasad, Nina Miolane, and Khanh Dao Duc. 2023. <span>“Using a Riemannian Elastic Metric for Statistical Analysis of Tumor Cell Shape Heterogeneity.”</span> In <em>International Conference on Geometric Science of Information</em>, 583–92. Springer.
</div>
<div id="ref-li2024unveiling" class="csl-entry" role="listitem">
———. 2024. <span>“Unveiling Cellular Morphology: Statistical Analysis Using a Riemannian Elastic Metric in Cancer Cell Image Datasets.”</span> <em>Information Geometry</em>, 1–15.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/bioshape-analysis\.github\.io\/blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>