<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Pavel Buklemishev">
<meta name="dcterms.date" content="2024-12-20">

<title>Exploring the relation between cell shape and motility – bioshape-analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-5027bf1c1f92ac6615724d89c8213d6a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">bioshape-analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Exploring the relation between cell shape and motility</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">Cell Morphology</div>
                <div class="quarto-category">Cell Migration</div>
                <div class="quarto-category">Differential Geometry</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Pavel Buklemishev </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">December 20, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#cell-spatial-migration" id="toc-cell-spatial-migration" class="nav-link active" data-scroll-target="#cell-spatial-migration">1. Cell Spatial Migration</a>
  <ul class="collapse">
  <li><a href="#modes-of-migration" id="toc-modes-of-migration" class="nav-link" data-scroll-target="#modes-of-migration">Modes of migration</a></li>
  <li><a href="#trajectories-visualization" id="toc-trajectories-visualization" class="nav-link" data-scroll-target="#trajectories-visualization">Trajectories visualization</a></li>
  <li><a href="#trajectory-classification." id="toc-trajectory-classification." class="nav-link" data-scroll-target="#trajectory-classification.">Trajectory classification.</a></li>
  <li><a href="#sec-mean_velo" id="toc-sec-mean_velo" class="nav-link" data-scroll-target="#sec-mean_velo">Dataset Motion Modes Analysis</a>
  <ul class="collapse">
  <li><a href="#motion-modes-distribution" id="toc-motion-modes-distribution" class="nav-link" data-scroll-target="#motion-modes-distribution">Motion Modes Distribution</a></li>
  <li><a href="#transition-between-modes-matrix" id="toc-transition-between-modes-matrix" class="nav-link" data-scroll-target="#transition-between-modes-matrix">Transition between Modes Matrix</a></li>
  <li><a href="#sec-avvel" id="toc-sec-avvel" class="nav-link" data-scroll-target="#sec-avvel">Free and Directed Diffusion: average velocities</a></li>
  <li><a href="#sec-avang" id="toc-sec-avang" class="nav-link" data-scroll-target="#sec-avang">Free and Directed Diffusion: average angles</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#shape-analysis" id="toc-shape-analysis" class="nav-link" data-scroll-target="#shape-analysis">2. Shape analysis</a>
  <ul class="collapse">
  <li><a href="#distance-computation" id="toc-distance-computation" class="nav-link" data-scroll-target="#distance-computation">Distance Computation</a></li>
  <li><a href="#free-and-directed-diffusion-riemann-velocities" id="toc-free-and-directed-diffusion-riemann-velocities" class="nav-link" data-scroll-target="#free-and-directed-diffusion-riemann-velocities">Free and Directed Diffusion: Riemann Velocities</a></li>
  </ul></li>
  <li><a href="#conformal-mapping" id="toc-conformal-mapping" class="nav-link" data-scroll-target="#conformal-mapping">Conformal mapping</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#future-plans" id="toc-future-plans" class="nav-link" data-scroll-target="#future-plans">Future plans</a></li>
  <li><a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix">Appendix</a>
  <ul class="collapse">
  <li><a href="#sec-organization" id="toc-sec-organization" class="nav-link" data-scroll-target="#sec-organization">Dataset organization</a></li>
  <li><a href="#sec-dcmssprep" id="toc-sec-dcmssprep" class="nav-link" data-scroll-target="#sec-dcmssprep">DC-MSS data preparation</a></li>
  <li><a href="#sec-align" id="toc-sec-align" class="nav-link" data-scroll-target="#sec-align">Alignment</a></li>
  <li><a href="#sec-riemann_simple" id="toc-sec-riemann_simple" class="nav-link" data-scroll-target="#sec-riemann_simple">Plotting Riemann Velocities of time</a></li>
  <li><a href="#sec-riemann_segmentation" id="toc-sec-riemann_segmentation" class="nav-link" data-scroll-target="#sec-riemann_segmentation">Plotting Riemann Distances with segmentation</a></li>
  <li><a href="#sec-someconf" id="toc-sec-someconf" class="nav-link" data-scroll-target="#sec-someconf">Additional examples: cell №87 topographical representations over migration modes transitions</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>In our previous proposal, we introduce our project to study how changes in cell shapes can be associated with motility and introduced our dataset. Here we report the methods and results we obtained in two main parts:</p>
<ol type="1">
<li><p><strong>Cell Spatial Migration</strong>:</p>
<p>We visualize and classify the trajectories of motion for cell dataset.</p></li>
<li><p><strong>Shape Dynamics</strong>:</p>
<p>We compute the Riemann distances over time and investigate characteristic events to understand cell shape behavior during migration. We also use conformal mapping to analyze protrusions and other shape features.</p></li>
</ol>
<section id="cell-spatial-migration" class="level1">
<h1>1. Cell Spatial Migration</h1>
<section id="modes-of-migration" class="level3">
<h3 class="anchored" data-anchor-id="modes-of-migration">Modes of migration</h3>
<p>Cells exhibit different migration patterns, such as free diffusion, directed migration, and confined motion. Mean Squared Displacement (MSD) analysis was used to distinguish these modes. <span class="citation" data-cites="Modes">(<a href="#ref-Modes" role="doc-biblioref">Dieterich et al. 2008</a>)</span> <span class="citation" data-cites="MSD">(<a href="#ref-MSD" role="doc-biblioref">Rosen 2021</a>)</span> <span class="citation" data-cites="wikipedia_anomalous_diffusion">(<a href="#ref-wikipedia_anomalous_diffusion" role="doc-biblioref">Wikipedia, n.d.</a>)</span>.</p>
<p>Mean squared displacement (MSD):</p>
<p><span class="math display">\[msd(t) = &lt;[x(t+t_0) - x(t)]^2 + [y(t+t_0) - y(t)]^2&gt; \]</span></p>
<p>Experimentally, the MSD depends on time in a polynomial way: <span class="math display">\[ msd(t)  = C t^{\alpha}\]</span></p>
<p>The motion types are described by the value of the parameter <span class="math inline">\(\alpha\)</span>(<a href="#fig-types" class="quarto-xref">Figure&nbsp;1</a>):</p>
<ul>
<li><strong><span class="math inline">\(\alpha = 1\)</span></strong>: Free Difusion.<br>
</li>
<li><strong><span class="math inline">\(\alpha = 2\)</span></strong>: Directed Diffusion.<br>
</li>
<li><strong><span class="math inline">\(1 &lt; \alpha &lt; 2\)</span></strong>: Superdiffusion.</li>
<li><strong><span class="math inline">\(\alpha &lt;1\)</span></strong>: Subdiffusion (anomalous diffusion, confined diffusion).</li>
<li><strong><span class="math inline">\(\alpha \approx 0\)</span></strong>: Immobility</li>
</ul>
<div id="fig-types" class="quarto-float quarto-figure quarto-figure-center anchored" style="display: block; margin-left: auto; margin-right: auto;">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-types-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/diff_types.png" class="img-fluid figure-img" style="display: block; margin-left: auto; margin-right: auto;;width:65.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-types-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Diffusion types <span class="citation" data-cites="picture">(<a href="#ref-picture" role="doc-biblioref">Kapanidis, Uphoff, and Stracy 2018</a>)</span>
</figcaption>
</figure>
</div>
<p>Using the trajectories, we aim to determine the motions type that appears in the trajectory of a cell and identify potential transitions between them. We discuss the trajectories analysis in the following sections.</p>
</section>
<section id="trajectories-visualization" class="level3">
<h3 class="anchored" data-anchor-id="trajectories-visualization">Trajectories visualization</h3>
<p>The cell can exhibit different diffusion behavior while migrating in space, thus, initially to investigate these modes of migration we need to visualize the cell trajectories. We visualize the cell trajectory using the following Python code(the dataset was initially adjusted <a href="#sec-organization" class="quarto-xref">Section&nbsp;6.0.1</a>):</p>
<section id="code" class="level4">
<h4 class="anchored" data-anchor-id="code">Code</h4>
<div id="3c7ebbb7" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_cell_trajectory(n):</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    centroids <span class="op">=</span> centr[n <span class="op">-</span> <span class="dv">1</span>]  </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    centroids <span class="op">=</span> np.array(centroids)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    x_coords <span class="op">=</span> centroids[:, <span class="dv">0</span>]  </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    y_coords <span class="op">=</span> centroids[:, <span class="dv">1</span>] </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># riemann_data = riemann[n-1] </span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># riemann_data = np.array(riemann_data)  </span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    time_steps <span class="op">=</span> np.arange(<span class="bu">len</span>(x_coords))  </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    plt.scatter(</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>        x_coords[<span class="dv">0</span>], y_coords[<span class="dv">0</span>],  </span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>        c<span class="op">=</span><span class="st">'black'</span>,</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>        marker<span class="op">=</span><span class="st">'o'</span>,</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>        edgecolor<span class="op">=</span><span class="st">'k'</span>,</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>        s<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.7</span>,</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>        label<span class="op">=</span><span class="st">'Start Point'</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    scatter <span class="op">=</span> plt.scatter(</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>        x_coords[<span class="dv">1</span>:], y_coords[<span class="dv">1</span>:],  </span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>        c<span class="op">=</span>time_steps[<span class="dv">1</span>:],  <span class="co">#riemann_data[1:],        </span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">'plasma'</span>,             </span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>        marker<span class="op">=</span><span class="st">'o'</span>,</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>        edgecolor<span class="op">=</span><span class="st">'k'</span>,</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>        s<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.7</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>    plt.plot(x_coords, y_coords, linestyle<span class="op">=</span><span class="st">'-'</span>, color<span class="op">=</span><span class="st">'gray'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)  </span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="ss">f"Cell Num </span><span class="sc">{</span>n<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">'X Coordinate'</span>)</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">'Y Coordinate'</span>)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>    plt.grid(<span class="va">True</span>)</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>    cbar <span class="op">=</span> plt.colorbar(scatter)</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>    cbar.set_label(<span class="st">'Time Step (t)'</span>, rotation<span class="op">=</span><span class="dv">270</span>, labelpad<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="fig-traj87" class="quarto-float quarto-figure quarto-figure-center anchored" style="display: block; margin-left: auto; margin-right: auto;">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-traj87-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/cell_traj_87.png" class="img-fluid figure-img" style="display: block; margin-left: auto; margin-right: auto;;width:65.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-traj87-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: The cell №87 trajectory
</figcaption>
</figure>
</div>
<p>We build the trajectories of the cells of the whole dataset. In the example figure(<a href="#fig-traj87" class="quarto-xref">Figure&nbsp;2</a>) one can see the trajectory of the cell №87. However, building trajectories itself doesn’t give an answer which modes a cell exhibits. Thus, we need to formulate and demonstrate a technique that help as to segmentize and classify the modes.</p>
<p><a href="files/trajectories.pdf">All the trajectories pdf</a></p>
</section>
</section>
<section id="trajectory-classification." class="level3">
<h3 class="anchored" data-anchor-id="trajectory-classification.">Trajectory classification.</h3>
<p><strong>DC-MSS</strong> - divide-and-conquer moment scaling spectrum transient mobility analysis framework <span class="citation" data-cites="VEGA20181018">(<a href="#ref-VEGA20181018" role="doc-biblioref">Vega et al. 2018</a>)</span> which classifies trajectory segments into predefined motion types and detects mobility switches. In our project, we use this tool to describe the trajectories of cells. Although this method is powerful and efficient for classifying trajectories, it has its cons as well. As it was designated for the molecular motion analysis it doesn’t take into account superdiffusion mode which is more typical for living species. Thus, method distinguishes between free diffusion, confined motion, immobility, and directed migration, and we will assume that if a cell exhibits superdiffusion mode, it will be covered by the lattest one.</p>
<p>After converting data to the framework format we can get desired mode classification.</p>
<section id="running-the-dc-mss-framework-in-matlab" class="level4">
<h4 class="anchored" data-anchor-id="running-the-dc-mss-framework-in-matlab">Running the DC-MSS Framework in MATLAB</h4>
<p>After preparing data for the framework(<a href="#sec-dcmssprep" class="quarto-xref">Section&nbsp;6.0.2</a>) we process the classification.</p>
<div id="9937b974" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>loaded <span class="op">=</span> load(<span class="st">"../trajectory_data.mat"</span>)<span class="op">;</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>allTracks <span class="op">=</span> loaded.tracks<span class="op">;</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>probDim <span class="op">=</span> <span class="dv">2</span><span class="op">;</span>       </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>plotRes <span class="op">=</span> <span class="dv">0</span><span class="op">;</span>       </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>peakAlpha <span class="op">=</span> <span class="dv">95</span><span class="op">;</span>  </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> struct()<span class="op">;</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>trackNames <span class="op">=</span> fieldnames(allTracks)<span class="op">;</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="op">=</span> <span class="dv">1</span>:length(trackNames)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    trackName <span class="op">=</span> trackNames{i}<span class="op">;</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    tracks <span class="op">=</span> allTracks.(trackName)<span class="op">;</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    [transDiffResults, errFlag] <span class="op">=</span> basicTransientDiffusionAnalysisv1(tracks, probDim, plotRes, peakAlpha)<span class="op">;</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> isfield(transDiffResults.segmentClass, <span class="st">'momentScalingSpectrum'</span>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>        results.(trackName).momentScalingSpectrum <span class="op">=</span> transDiffResults.segmentClass.momentScalingSpectrum<span class="op">;</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    end</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>end</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>h5FileName <span class="op">=</span> <span class="st">'time_events.h5'</span><span class="op">;</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>trackNames <span class="op">=</span> fieldnames(results)<span class="op">;</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="op">=</span> <span class="dv">1</span>:length(trackNames)</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    trackName <span class="op">=</span> trackNames{i}<span class="op">;</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>    data <span class="op">=</span> results.(trackName).momentScalingSpectrum<span class="op">;</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>end</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>By varying <strong>peakAlpha</strong> parameter one can get different confidence level for choosing peaks when initially segmenting track.</p>
<div class="row">
<div class="column" style="width:40%;">
<div id="fig-95-class" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-95-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/segmented_track.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-95-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Classified track with 95% confidence level. Cyan corresponds to the free diffusion, Magenta - directed motion
</figcaption>
</figure>
</div>
</div>
<div class="column" style="width:47%;">
<div id="fig-90-class" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-90-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/conf_level_90.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-90-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Classified track with 90% confidence level. Blue corresponds to the confined diffusion
</figcaption>
</figure>
</div>
</div>
</div>
<p>In the example (<a href="#fig-95-class" class="quarto-xref">Figure&nbsp;3</a> and <a href="#fig-90-class" class="quarto-xref">Figure&nbsp;4</a>) we can see that the classified trajectory segments provide insights into motion types. However, (<a href="#fig-90-class" class="quarto-xref">Figure&nbsp;4</a>) looks much more natively as the motion changes in the cyan part of trajectory(<a href="#fig-95-class" class="quarto-xref">Figure&nbsp;3</a>) are visible to the naked eye.</p>
</section>
</section>
<section id="sec-mean_velo" class="level2">
<h2 class="anchored" data-anchor-id="sec-mean_velo">Dataset Motion Modes Analysis</h2>
<p>In this section, we organize the dataset based on the migration modes classification discussed in the previous sections.</p>
<p>First, we will estimate the proportion of each observed migration mode relative to the total.</p>
<section id="motion-modes-distribution" class="level3">
<h3 class="anchored" data-anchor-id="motion-modes-distribution">Motion Modes Distribution</h3>
<p>Using the illustrations (<a href="#fig-prop95" class="quarto-xref">Figure&nbsp;5</a>, <a href="#fig-prop90" class="quarto-xref">Figure&nbsp;6</a>), it is evident that, despite the presence of unclassified trajectory segments, the two predominant regimes describing the trajectories are the free and confined migration modes.</p>
<div id="c94390d9" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_motion_type_distribution():</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    type_counts <span class="op">=</span> defaultdict(<span class="bu">int</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    total_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_95.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> plot_index <span class="kw">in</span> <span class="bu">range</span>(total_plots):</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>            track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>plot_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>            first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>            interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> interval_type <span class="kw">in</span> interval_types:</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="kw">not</span> np.isnan(interval_type):</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>                    type_counts[<span class="bu">int</span>(interval_type)] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>                    total_count <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                    type_counts[<span class="st">"unclassified"</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>                    total_count <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    proportions <span class="op">=</span> {key: count <span class="op">/</span> total_count <span class="cf">for</span> key, count <span class="kw">in</span> type_counts.items()}</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> proportions</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_motion_type_distribution():</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>    proportions <span class="op">=</span> get_motion_type_distribution()</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>    types <span class="op">=</span> [<span class="st">"Immobile"</span>, <span class="st">"Confined Diffusion"</span>, <span class="st">"Free Diffusion"</span>, <span class="st">"Directed Diffusion"</span>, <span class="st">"Unclassified"</span>]</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">"brown"</span>, <span class="st">"blue"</span>, <span class="st">"cyan"</span>, <span class="st">"magenta"</span>, <span class="st">"black"</span>]</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    keys <span class="op">=</span> [<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="st">"unclassified"</span>]</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>    values <span class="op">=</span> [proportions.get(key, <span class="dv">0</span>) <span class="cf">for</span> key <span class="kw">in</span> keys]</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>    plt.pie(</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>        values,</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>        labels<span class="op">=</span>types,</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>        colors<span class="op">=</span>colors,</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>        autopct<span class="op">=</span><span class="st">"</span><span class="sc">%1.1f%%</span><span class="st">"</span>,</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>        startangle<span class="op">=</span><span class="dv">140</span>,</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>        textprops<span class="op">=</span>{<span class="st">'color'</span>: <span class="st">'white'</span>, <span class="st">'fontsize'</span>: <span class="dv">12</span>}  </span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">"Proportion of Motion Types"</span>)</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">"motion_type_distribution_90.png"</span>)</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-prop95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-prop95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/prop95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-prop95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Confidence level 0.95. Cyan corresponds to the free diffusion, Magenta - directed motion, Blue - confined diffusion, Brown - to immobile cells, Black - unclassified.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-prop90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-prop90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/prop90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-prop90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="transition-between-modes-matrix" class="level3">
<h3 class="anchored" data-anchor-id="transition-between-modes-matrix">Transition between Modes Matrix</h3>
<p>The transition fractions of mode switches can also be represented using a Transition Matrix. These matrices (<a href="#fig-transmat95" class="quarto-xref">Figure&nbsp;7</a>, <a href="#fig-transmat90" class="quarto-xref">Figure&nbsp;8</a>) illustrate the most frequent transitions between migration modes, offering valuable insight into the dynamics of mode switching. From these illustrations, we can draw the following conclusions:</p>
<ol type="1">
<li><p>Probabilities of the transition from one mode to another usually depends on the direction of a transition.</p></li>
<li><p>The most dominant mode transition is from free diffusion to the confined mode.</p></li>
<li><p>Cells tend to switch between modes in a sequential manner, following the pattern: Immobile → Confined → Free → Directed and in the opposite direction.</p></li>
</ol>
<div id="e75770c1" class="cell" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> compute_transition_matrix_with_unclassified():</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    transition_counts <span class="op">=</span> np.zeros((<span class="dv">5</span>, <span class="dv">5</span>))  </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    total_transitions <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_95.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> plot_index <span class="kw">in</span> <span class="bu">range</span>(total_plots):</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>            track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>plot_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>            first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>            interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> start_idx <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(interval_types) <span class="op">-</span> <span class="dv">1</span>):</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                type_from <span class="op">=</span> interval_types[start_idx]</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                type_to <span class="op">=</span> interval_types[start_idx <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>                type_from <span class="op">=</span> <span class="bu">int</span>(type_from) <span class="cf">if</span> <span class="kw">not</span> np.isnan(type_from) <span class="cf">else</span> <span class="dv">4</span>  </span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>                type_to <span class="op">=</span> <span class="bu">int</span>(type_to) <span class="cf">if</span> <span class="kw">not</span> np.isnan(type_to) <span class="cf">else</span> <span class="dv">4</span>  </span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> type_from <span class="op">!=</span> type_to:  </span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>                    transition_counts[type_from, type_to] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>                    total_transitions <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>    transition_matrix <span class="op">=</span> (transition_counts <span class="op">/</span> total_transitions) <span class="op">*</span> <span class="dv">100</span>  </span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> transition_matrix</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_transition_matrix_with_unclassified(transition_matrix):</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>    types <span class="op">=</span> [<span class="st">"Immobile"</span>, <span class="st">"Confined Diffusion"</span>, <span class="st">"Free Diffusion"</span>, <span class="st">"Directed Diffusion"</span>, <span class="st">"Unclassified"</span>]</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>    masked_matrix <span class="op">=</span> np.ma.masked_where(np.eye(<span class="bu">len</span>(types)), transition_matrix)  </span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>    plt.imshow(masked_matrix, cmap<span class="op">=</span><span class="st">"Blues"</span>, aspect<span class="op">=</span><span class="st">"auto"</span>)</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">5</span>):</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>        plt.fill([i <span class="op">-</span> <span class="fl">0.5</span>, i <span class="op">+</span> <span class="fl">0.5</span>, i <span class="op">+</span> <span class="fl">0.5</span>, i <span class="op">-</span> <span class="fl">0.5</span>],</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>                 [i <span class="op">-</span> <span class="fl">0.5</span>, i <span class="op">-</span> <span class="fl">0.5</span>, i <span class="op">+</span> <span class="fl">0.5</span>, i <span class="op">+</span> <span class="fl">0.5</span>],</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>                 color<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>    plt.colorbar(label<span class="op">=</span><span class="st">"%"</span>)</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>    plt.xticks(ticks<span class="op">=</span>np.arange(<span class="bu">len</span>(types)), labels<span class="op">=</span>types, rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>    plt.yticks(ticks<span class="op">=</span>np.arange(<span class="bu">len</span>(types)), labels<span class="op">=</span>types)</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">"Transition Matrix"</span>)</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(types)):</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(types)):</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i <span class="op">!=</span> j:  </span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>                plt.text(j, i, <span class="ss">f"</span><span class="sc">{</span>transition_matrix[i, j]<span class="sc">:.1f}</span><span class="ss">%"</span>, ha<span class="op">=</span><span class="st">'center'</span>, va<span class="op">=</span><span class="st">'center'</span>, color<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">"transition_matrix_with_unclassified_95.png"</span>)</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-transmat95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-transmat95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/transition_matrix_with_unclassified_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-transmat95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Confidence level 0.95. The Y-axis represents the migration mode from which cells transition to a new observed mode, while the X-axis corresponds to the mode they transition into.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-transmat90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-transmat90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/transition_matrix_with_unclassified_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-transmat90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-avvel" class="level3">
<h3 class="anchored" data-anchor-id="sec-avvel">Free and Directed Diffusion: average velocities</h3>
<p>Since the most interpretable regimes for observers are free diffusion and directed motion, in the following subsections, we investigate cell motility characteristics concerning these two modes. We compute the mean velocities associated with these migration modes, presenting both the mean values and their 95% confidence intervals. Unfortunately, the p-value for the t-test at the 95% confidence level segmentation(<a href="#fig-meanvelocity95" class="quarto-xref">Figure&nbsp;9</a>) exceeds the required 0.05, which limits the statistical significance of the observed differences. However, the segmentation at the 90% confidence level demonstrates statistical significance for the observed velocities. As we can see in (<a href="#fig-meanvelocity90" class="quarto-xref">Figure&nbsp;10</a>), the velocity in directed migration is higher than in free diffusion. This can be explained by the active mechanisms involved in directed migration, such as the reorganization of the cytoskeleton, enhanced energy expenditure, which result in more efficient and faster movement compared to the random free diffusion.</p>
<div id="34b0a208" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> average_velocity_with_ttest():</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    type_data <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_95.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> plot_index <span class="kw">in</span> <span class="bu">range</span>(total_plots):</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>            track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>plot_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>            first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>            event_indices <span class="op">=</span> first_three_rows[<span class="dv">0</span>, :].astype(<span class="bu">int</span>) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>            interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> start_idx, interval_type <span class="kw">in</span> <span class="bu">enumerate</span>(interval_types):</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> start_idx <span class="op">+</span> <span class="dv">1</span> <span class="op">&lt;</span> <span class="bu">len</span>(event_indices) <span class="kw">and</span> interval_type <span class="kw">in</span> [<span class="dv">2</span>, <span class="dv">3</span>]:                      start <span class="op">=</span> event_indices[start_idx] <span class="op">+</span> <span class="dv">1</span>  </span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>                    end <span class="op">=</span> event_indices[start_idx <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>                    segment <span class="op">=</span> [get_abs_velocity(plot_index, idx) <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(start, end <span class="op">+</span> <span class="dv">1</span>)]</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>                    type_data[<span class="bu">int</span>(interval_type)].extend(segment)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    free_diffusion <span class="op">=</span> type_data[<span class="dv">2</span>]</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    directed_diffusion <span class="op">=</span> type_data[<span class="dv">3</span>]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    t_stat, p_value <span class="op">=</span> ttest_ind(free_diffusion, directed_diffusion, equal_var<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    types <span class="op">=</span> [<span class="st">"Free Diffusion"</span>, <span class="st">"Directed Diffusion"</span>]</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">"cyan"</span>, <span class="st">"magenta"</span>]</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>    means <span class="op">=</span> []</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>    conf_intervals <span class="op">=</span> []</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> key, velocities <span class="kw">in</span> <span class="bu">zip</span>([<span class="dv">2</span>, <span class="dv">3</span>], [free_diffusion, directed_diffusion]):</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> velocities:  </span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>            mean <span class="op">=</span> np.mean(velocities)</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>            ci <span class="op">=</span> sem(velocities) <span class="op">*</span> <span class="fl">1.96</span>  </span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>            means.append(mean)</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>            conf_intervals.append(ci)</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>            means.append(<span class="dv">0</span>)</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>            conf_intervals.append(<span class="dv">0</span>)</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> np.arange(<span class="bu">len</span>(types))</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> idx, (mean, ci, color) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(means, conf_intervals, colors)):</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>        plt.errorbar(x[idx], mean, yerr<span class="op">=</span>ci, fmt<span class="op">=</span><span class="st">'o'</span>, color<span class="op">=</span>color, ecolor<span class="op">=</span>color, elinewidth<span class="op">=</span><span class="dv">2</span>, capsize<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>    plt.text(<span class="fl">0.5</span>, <span class="bu">max</span>(means) <span class="op">+</span> <span class="bu">max</span>(conf_intervals) <span class="op">*</span> <span class="fl">1.2</span>,</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>             <span class="ss">f"T-test p-value: </span><span class="sc">{</span>p_value<span class="sc">:.3e}</span><span class="ss">"</span>,</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>             ha<span class="op">=</span><span class="st">'center'</span>, fontsize<span class="op">=</span><span class="dv">12</span>, color<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>    plt.xticks(x, types)</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">"Velocity"</span>)</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">"mean_velocity_with_ttest_95.png"</span>)</span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-meanvelocity95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-meanvelocity95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/mean_velocity_with_ttest_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-meanvelocity95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-meanvelocity90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-meanvelocity90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/mean_velocity_with_ttest_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-meanvelocity90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-avang" class="level3">
<h3 class="anchored" data-anchor-id="sec-avang">Free and Directed Diffusion: average angles</h3>
<p>Analogously to the previous subsection, we measured the average directional angles associated with different migration modes. Both segmentation levels satisfy the p-value criterion, confirming the statistical significance of the results. From both images, it is evident that cells exhibit more stable behavior in the directed diffusion mode, with noticeably less rotational movement.</p>
<div id="b5896063" class="cell" data-execution_count="6">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> average_angle_with_ttest():</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    type_data <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_90.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> plot_index <span class="kw">in</span> <span class="bu">range</span>(total_plots):</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>            track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>plot_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>            first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>            event_indices <span class="op">=</span> first_three_rows[<span class="dv">0</span>, :].astype(<span class="bu">int</span>) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>            interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> start_idx, interval_type <span class="kw">in</span> <span class="bu">enumerate</span>(interval_types):</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> start_idx <span class="op">+</span> <span class="dv">1</span> <span class="op">&lt;</span> <span class="bu">len</span>(event_indices) <span class="kw">and</span> interval_type <span class="kw">in</span> [<span class="dv">2</span>, <span class="dv">3</span>]:                      start <span class="op">=</span> event_indices[start_idx] <span class="op">+</span> <span class="dv">1</span> </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>                    end <span class="op">=</span> event_indices[start_idx <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                    segment <span class="op">=</span> [get_velocity_angle_rel(plot_index, idx) <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(start, end <span class="op">+</span> <span class="dv">1</span>)]</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>                    type_data[<span class="bu">int</span>(interval_type)].extend(segment)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    free_diffusion <span class="op">=</span> type_data[<span class="dv">2</span>]</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    directed_diffusion <span class="op">=</span> type_data[<span class="dv">3</span>]</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    t_stat, p_value <span class="op">=</span> ttest_ind(free_diffusion, directed_diffusion, equal_var<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>    types <span class="op">=</span> [<span class="st">"Free Diffusion"</span>, <span class="st">"Directed Diffusion"</span>]</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">"cyan"</span>, <span class="st">"magenta"</span>]</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>    means <span class="op">=</span> []</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>    conf_intervals <span class="op">=</span> []</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> key, angles <span class="kw">in</span> <span class="bu">zip</span>([<span class="dv">2</span>, <span class="dv">3</span>], [free_diffusion, directed_diffusion]):</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> angles: </span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>            sigma <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>            angles_smoothed <span class="op">=</span> gaussian_filter1d(angles, sigma<span class="op">=</span>sigma)</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a>            mean <span class="op">=</span> np.mean(angles_smoothed)</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a>            ci <span class="op">=</span> sem(angles_smoothed) <span class="op">*</span> <span class="fl">1.96</span> </span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>            means.append(mean)</span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>            conf_intervals.append(ci)</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>            means.append(<span class="dv">0</span>)</span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a>            conf_intervals.append(<span class="dv">0</span>)</span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> np.arange(<span class="bu">len</span>(types))</span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> idx, (mean, ci, color) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(means, conf_intervals, colors)):</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true" tabindex="-1"></a>        plt.errorbar(x[idx], mean, yerr<span class="op">=</span>ci, fmt<span class="op">=</span><span class="st">'o'</span>, color<span class="op">=</span>color, ecolor<span class="op">=</span>color, elinewidth<span class="op">=</span><span class="dv">2</span>, capsize<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true" tabindex="-1"></a>    plt.text(<span class="fl">0.5</span>, <span class="bu">max</span>(means) <span class="op">+</span> <span class="bu">max</span>(conf_intervals) <span class="op">*</span> <span class="fl">1.2</span>,</span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true" tabindex="-1"></a>             <span class="ss">f"T-test p-value: </span><span class="sc">{</span>p_value<span class="sc">:.3e}</span><span class="ss">"</span>,</span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true" tabindex="-1"></a>             ha<span class="op">=</span><span class="st">'center'</span>, fontsize<span class="op">=</span><span class="dv">12</span>, color<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb6-52"><a href="#cb6-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-53"><a href="#cb6-53" aria-hidden="true" tabindex="-1"></a>    plt.xticks(x, types)</span>
<span id="cb6-54"><a href="#cb6-54" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">"Angle"</span>)</span>
<span id="cb6-55"><a href="#cb6-55" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb6-56"><a href="#cb6-56" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">"mean_angle_with_ttest_90.png"</span>)</span>
<span id="cb6-57"><a href="#cb6-57" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-meanangle95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-meanangle95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/mean_angle_with_ttest_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-meanangle95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-meanangle90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-meanangle90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/mean_angle_with_ttest_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-meanangle90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;12: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="shape-analysis" class="level1">
<h1>2. Shape analysis</h1>
<p>After describing the spatial migration of a cell, our next goal is to investigate the dynamics of its shape. We employ the SRV Riemann Elastic Metrics to compute the distance between two shapes<span class="citation" data-cites="10.1007/978-3-031-38299-4_60">(<a href="#ref-10.1007/978-3-031-38299-4_60" role="doc-biblioref">Li et al. 2023</a>)</span>:</p>
<p><span class="math display">\[
g_c^{1, 0.5}(h, k) = \int_{[0,1]} \langle D_s h, N \rangle \langle D_s k, N \rangle \, ds
+ 0.5^2 \int_{[0,1]} \langle D_s h, T \rangle \langle D_s k, T \rangle \, ds  \text{ - SRV metrics}
\]</span></p>
<p>For a single cell given time moments <span class="math inline">\(t\)</span> and <span class="math inline">\(t+1\)</span>, we:</p>
<ol type="1">
<li><p>Interpolate the segmentized cell shape for both of moments t and t+1</p></li>
<li><p>Align the cell shapes. We quotient out transition and reparametrization but ignore the rotation. We expect that the rotation is a significant aspect in spatial migration and we need to focus it.</p></li>
<li><p>Compute the SRV metrics for an each couple of consequent aligned cell shapes(<a href="#sec-align" class="quarto-xref">Section&nbsp;6.0.3</a>), dividing them by time differences, and, therefore we obtain Riemann Velocities (They would be mentioned in the text as distances as well).</p></li>
</ol>
<section id="distance-computation" class="level3">
<h3 class="anchored" data-anchor-id="distance-computation">Distance Computation</h3>
<p>Riemann distances are computed between consequent aligned shapes:</p>
<div id="0844fcb5" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>riemann_distances <span class="op">=</span> []</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>a <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>b <span class="op">=</span> <span class="dv">1</span><span class="op">/</span><span class="dv">2</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>CURVES_SPACE_ELASTIC <span class="op">=</span> DiscreteCurvesStartingAtOrigin(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    ambient_dim<span class="op">=</span><span class="dv">2</span>, k_sampling_points<span class="op">=</span><span class="dv">1000</span>, equip<span class="op">=</span><span class="va">False</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>CURVES_SPACE_ELASTIC.equip_with_metric(ElasticMetric, a<span class="op">=</span>a, b<span class="op">=</span>b)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calculate_distance(border,reference_shape):</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> CURVES_SPACE_ELASTIC.metric.dist(CURVES_SPACE_ELASTIC.projection(border), CURVES_SPACE_ELASTIC.projection(reference_shape))</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> cell_i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="dv">205</span>):</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    number_of_frames <span class="op">=</span> <span class="bu">sum</span>(os.path.isdir(os.path.join(<span class="ss">f"cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">"</span>, entry)) <span class="cf">for</span> entry <span class="kw">in</span> os.listdir(<span class="ss">f"cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">"</span>))  </span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    iter_distance <span class="op">=</span> np.zeros(number_of_frames)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>    BASE_LINE <span class="op">=</span> np.load(<span class="ss">f'cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">/frame_1/outline.npy'</span>)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    BASE_LINE<span class="op">=</span> interpolate(BASE_LINE,<span class="dv">1000</span>)</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    BASE_LINE <span class="op">=</span> preprocess(BASE_LINE)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>    BASE_LINE<span class="op">=</span> project_on_kendall_space(BASE_LINE)</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(number_of_frames):</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>        border_cell <span class="op">=</span> np.load(<span class="ss">f'cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">/frame_</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">/outline.npy'</span>)</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>        cell_interpolation<span class="op">=</span> interpolate(border_cell,<span class="dv">1000</span>)</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>        cell_preprocess <span class="op">=</span> preprocess(cell_interpolation)</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>        border_cell <span class="op">=</span> cell_preprocess</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>        border_cell <span class="op">=</span> project_on_kendall_space(cell_interpolation)</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>        aligned_border <span class="op">=</span> align(border_cell, BASE_LINE, rescale<span class="op">=</span><span class="va">True</span>, rotation<span class="op">=</span><span class="va">False</span>, reparameterization<span class="op">=</span><span class="va">True</span>, k_sampling_points<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>        iter_distance[i] <span class="op">=</span> calculate_distance(aligned_border, BASE_LINE)</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>        BASE_LINE <span class="op">=</span> aligned_border </span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>    riemann_distances.append(iter_distance)</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="co">### Dividing by delta t in the results.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Riemann Velocities visualization - (<a href="#sec-riemann_simple" class="quarto-xref">Section&nbsp;6.0.4</a>).</p>
<p>Moreover, we hypothesize that the cells’ shape dynamics, as described by Riemann Velocities, are influenced by the migration mode they exhibit. Therefore, we align the Riemann Velocities with the mode segmentation obtained earlier for further analysis.</p>
<p>Riemann Velocities visualization with segmentation - (<a href="#sec-riemann_segmentation" class="quarto-xref">Section&nbsp;6.0.5</a>).</p>
<p>From the example pictures (<a href="#fig-segm095" class="quarto-xref">Figure&nbsp;13</a>, <a href="#fig-segm090" class="quarto-xref">Figure&nbsp;14</a>), the peak at time point 49 aligns well with the observed transition in migration modes, suggesting that a connection may exist.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-segm095" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-segm095-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/riemann_single_cell_87_classification_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-segm095-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;13: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-segm090" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-segm090-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/riemann_single_cell_87_classification_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-segm090-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;14: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Based on the general results, we can formulate a hypothesis: Riemann velocity peaks emerge when the motion mode switches between Directed Diffusion and Confined/Free Diffusion modes. However, with the current data, we cannot confidently confirm this correlation. Nevertheless, some examples (e.g., cells №33, 36, 41, 57, etc.) suggest that this dependency could be a promising target for future exploration.</p>
<p><a href="files/riemann_times.pdf">All the Riemann Velocities pdf</a></p>
<p><a href="files/riemann_with_events_90.pdf">Riemann velocities and events (0.9 confidence level)</a></p>
<p><a href="files/riemann_with_events_95.pdf">Riemann velocities and events (0.95 confidence level)</a></p>
<p><a href="files/riemann_with_classification_90.pdf">Riemann velocities and classified segments (0.9 confidence level)</a></p>
<p><a href="files/riemann_with_classification_95.pdf">Riemann velocities and classified segments (0.95 confidence level)</a></p>
<p>We can visualize the spatial behaviour of a cell via the motion types as well (<a href="#fig-3d95" class="quarto-xref">Figure&nbsp;15</a>, <a href="#fig-3d90" class="quarto-xref">Figure&nbsp;16</a>). This could be helpful in the further research.</p>
<div id="7636c6d6" class="cell" data-execution_count="8">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_cell_by_motion_type(n):</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    cell_dir <span class="op">=</span> <span class="st">'cells'</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    cell_path <span class="op">=</span> os.path.join(</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>        cell_dir, </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>        <span class="bu">sorted</span>(os.listdir(cell_dir), key<span class="op">=</span><span class="kw">lambda</span> x: <span class="bu">int</span>(x.split(<span class="st">'_'</span>)[<span class="dv">1</span>]) <span class="cf">if</span> <span class="st">'_'</span> <span class="kw">in</span> x <span class="kw">and</span> x.split(<span class="st">'_'</span>)[<span class="dv">1</span>].isdigit() <span class="cf">else</span> <span class="dv">0</span>)[n<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    cell <span class="op">=</span> <span class="bu">sorted</span>(</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>        os.listdir(cell_path), </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>        key<span class="op">=</span><span class="kw">lambda</span> x: <span class="bu">int</span>(<span class="st">''</span>.join(<span class="bu">filter</span>(<span class="bu">str</span>.isdigit, x))) </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    fig <span class="op">=</span> plt.figure()</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>, projection<span class="op">=</span><span class="st">'3d'</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    interval_colors <span class="op">=</span> {</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>        <span class="dv">0</span>: <span class="st">"brown"</span>, </span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>        <span class="dv">1</span>: <span class="st">"blue"</span>,       </span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>        <span class="dv">2</span>: <span class="st">"cyan"</span>,       </span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>        <span class="dv">3</span>: <span class="st">"magenta"</span>,  </span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>        <span class="st">"unclassified"</span>: <span class="st">"black"</span>  </span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> h5py</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_90.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>        track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>n<span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>        first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>        event_indices <span class="op">=</span> first_three_rows[<span class="dv">0</span>, :].astype(<span class="bu">int</span>) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>        interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]  </span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, frame <span class="kw">in</span> <span class="bu">enumerate</span>(cell):</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>        frame_path <span class="op">=</span> os.path.join(cell_path, frame)</span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>        time <span class="op">=</span> np.load(os.path.join(frame_path, <span class="st">'time.npy'</span>))</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>        outline <span class="op">=</span> np.load(os.path.join(frame_path, <span class="st">'outline.npy'</span>))</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>        current_type <span class="op">=</span> <span class="va">None</span></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> start_idx, interval_type <span class="kw">in</span> <span class="bu">enumerate</span>(interval_types):</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i <span class="op">&gt;=</span> event_indices[start_idx] <span class="kw">and</span> (start_idx <span class="op">+</span> <span class="dv">1</span> <span class="op">==</span> <span class="bu">len</span>(event_indices) <span class="kw">or</span> i <span class="op">&lt;</span> event_indices[start_idx <span class="op">+</span> <span class="dv">1</span>]):</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>                current_type <span class="op">=</span> interval_type</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>                <span class="cf">break</span></span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> current_type <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>:</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>            interval_type <span class="op">=</span> <span class="bu">int</span>(current_type) <span class="cf">if</span> <span class="kw">not</span> np.isnan(current_type) <span class="cf">else</span> <span class="st">"unclassified"</span></span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>            color <span class="op">=</span> interval_colors.get(interval_type, <span class="st">"black"</span>)</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>        ax.plot3D(</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>            outline[:, <span class="dv">0</span>], </span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>            outline[:, <span class="dv">1</span>], </span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>            np.full(<span class="bu">len</span>(outline[:, <span class="dv">1</span>]), time),</span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>            color<span class="op">=</span>color, </span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>            linewidth<span class="op">=</span><span class="dv">1</span></span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Frame </span><span class="sc">{</span>frame<span class="sc">}</span><span class="ss">: Time = </span><span class="sc">{</span>time<span class="sc">}</span><span class="ss">, Motion Type = </span><span class="sc">{</span>current_type<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> matplotlib.lines <span class="im">import</span> Line2D</span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a>    legend_elements <span class="op">=</span> [</span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"brown"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Immobile"</span>),</span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"blue"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Confined Diffusion"</span>),</span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"cyan"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Free Diffusion"</span>),</span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"magenta"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Directed Diffusion"</span>),</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"black"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Unclassified"</span>)</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>    ax.legend(handles<span class="op">=</span>legend_elements, loc<span class="op">=</span><span class="st">"upper right"</span>, fontsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a>    ax.set_xlabel(<span class="st">'X Coordinate'</span>)</span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a>    ax.set_ylabel(<span class="st">'Y Coordinate'</span>)</span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a>    ax.set_zlabel(<span class="st">'Time'</span>)</span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="ss">f"Cell </span><span class="sc">{</span>n<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-3d95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-3d95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/cell87_3d_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3d95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-3d90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-3d90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/cell87_3d_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3d90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;16: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="free-and-directed-diffusion-riemann-velocities" class="level3">
<h3 class="anchored" data-anchor-id="free-and-directed-diffusion-riemann-velocities">Free and Directed Diffusion: Riemann Velocities</h3>
<p>In this section, similar to (<a href="#sec-avvel" class="quarto-xref">Section&nbsp;1.1.3</a>, <a href="#sec-avang" class="quarto-xref">Section&nbsp;1.1.4</a>), we compute the mean values of Riemann Velocities associated with Free and Directed Diffusion modes.</p>
<p>The figures (<a href="#fig-ttestrim95" class="quarto-xref">Figure&nbsp;17</a>, <a href="#fig-ttestrim90" class="quarto-xref">Figure&nbsp;18</a>) illustrate that during directed migration, cells exhibit fewer shape perturbations compared to free diffusion modes. This suggests that the cytoskeleton in cells undergoing directed migration is more stable than in cells in diffusive modes.</p>
<div id="f2c19a51" class="cell" data-execution_count="9">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> average_riemann_distances_with_ttest():</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    type_data <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_90.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> plot_index <span class="kw">in</span> <span class="bu">range</span>(total_plots):</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>            track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>plot_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>            first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>            event_indices <span class="op">=</span> first_three_rows[<span class="dv">0</span>, :].astype(<span class="bu">int</span>) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>            interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> start_idx, interval_type <span class="kw">in</span> <span class="bu">enumerate</span>(interval_types):</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> start_idx <span class="op">+</span> <span class="dv">1</span> <span class="op">&lt;</span> <span class="bu">len</span>(event_indices) <span class="kw">and</span> interval_type <span class="kw">in</span> [<span class="dv">2</span>, <span class="dv">3</span>]: </span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>                    start <span class="op">=</span> event_indices[start_idx] <span class="op">+</span> <span class="dv">1</span>  </span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>                    end <span class="op">=</span> event_indices[start_idx <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>                    segment <span class="op">=</span> [</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>                        get_riemann_dist(plot_index, idx) <span class="op">/</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>                        (get_times(plot_index, idx) <span class="op">-</span> get_times(plot_index, idx <span class="op">-</span> <span class="dv">1</span>))</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(start, end <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>                    ]</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>                    type_data[<span class="bu">int</span>(interval_type)].extend(segment)</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>    free_diffusion <span class="op">=</span> type_data[<span class="dv">2</span>]</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>    directed_diffusion <span class="op">=</span> type_data[<span class="dv">3</span>]</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    t_stat, p_value <span class="op">=</span> ttest_ind(free_diffusion, directed_diffusion, equal_var<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>    types <span class="op">=</span> [<span class="st">"Free Diffusion"</span>, <span class="st">"Directed Diffusion"</span>]</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">"cyan"</span>, <span class="st">"magenta"</span>]</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>    means <span class="op">=</span> []</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    conf_intervals <span class="op">=</span> []</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> key, distances <span class="kw">in</span> <span class="bu">zip</span>([<span class="dv">2</span>, <span class="dv">3</span>], [free_diffusion, directed_diffusion]):</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> distances:  </span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>            mean <span class="op">=</span> np.mean(distances)</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>            ci <span class="op">=</span> sem(distances) <span class="op">*</span> <span class="fl">1.96</span>  </span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>            means.append(mean)</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>            conf_intervals.append(ci)</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>            means.append(<span class="dv">0</span>)</span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>            conf_intervals.append(<span class="dv">0</span>)</span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> np.arange(<span class="bu">len</span>(types))</span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> idx, (mean, ci, color) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(means, conf_intervals, colors)):</span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>        plt.errorbar(x[idx], mean, yerr<span class="op">=</span>ci, fmt<span class="op">=</span><span class="st">'o'</span>, color<span class="op">=</span>color, ecolor<span class="op">=</span>color, elinewidth<span class="op">=</span><span class="dv">2</span>, capsize<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a>    plt.text(<span class="fl">0.5</span>, <span class="bu">max</span>(means) <span class="op">+</span> <span class="bu">max</span>(conf_intervals) <span class="op">*</span> <span class="fl">1.2</span>,</span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a>             <span class="ss">f"T-test p-value: </span><span class="sc">{</span>p_value<span class="sc">:.3e}</span><span class="ss">"</span>,</span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a>             ha<span class="op">=</span><span class="st">'center'</span>, fontsize<span class="op">=</span><span class="dv">12</span>, color<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a>    plt.xticks(x, types)</span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">"Riemann Velocity"</span>)</span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-56"><a href="#cb9-56" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb9-57"><a href="#cb9-57" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">"riemann_distances_with_ttest_90.png"</span>)</span>
<span id="cb9-58"><a href="#cb9-58" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="row">
<div class="column" style="width:49%;">
<div id="fig-ttestrim95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ttestrim95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/riemann_distances_with_ttest_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ttestrim95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div>
<div class="column" style="width:49%;">
<div id="fig-ttestrim90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ttestrim90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/riemann_distances_with_ttest_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ttestrim90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18: Confidence level 0.90.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Continuing this investigation, we can create scatter plots to visualize the relationships between Riemann velocities and angles, as well as between Riemann velocities and distances. These plots (<a href="#fig-scat95" class="quarto-xref">Figure&nbsp;19</a>, <a href="#fig-scat90" class="quarto-xref">Figure&nbsp;20</a>) illustrate that in the directed migration mode, the scatter of Riemann distances is more concentrated, showing less variability in the Directed Migration mode.</p>
<div id="ed56f14d" class="cell" data-execution_count="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_velocity_riemann_and_angle_riemann_for_two_modes():</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    type_data <span class="op">=</span> defaultdict(<span class="kw">lambda</span>: {<span class="st">'velocity'</span>: [], <span class="st">'riemann'</span>: [], <span class="st">'angle'</span>: []})</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    colors <span class="op">=</span> {<span class="dv">2</span>: <span class="st">"cyan"</span>, <span class="dv">3</span>: <span class="st">"magenta"</span>}</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> {<span class="dv">2</span>: <span class="st">"Free Diffusion"</span>, <span class="dv">3</span>: <span class="st">"Directed Diffusion"</span>}</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_95.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> plot_index <span class="kw">in</span> <span class="bu">range</span>(total_plots):</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>            track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>plot_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>            first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>            event_indices <span class="op">=</span> first_three_rows[<span class="dv">0</span>, :].astype(<span class="bu">int</span>) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>            interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> start_idx, interval_type <span class="kw">in</span> <span class="bu">enumerate</span>(interval_types):</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> start_idx <span class="op">+</span> <span class="dv">1</span> <span class="op">&lt;</span> <span class="bu">len</span>(event_indices) <span class="kw">and</span> interval_type <span class="kw">in</span> [<span class="dv">2</span>, <span class="dv">3</span>]: </span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>                    start <span class="op">=</span> event_indices[start_idx] <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>                    end <span class="op">=</span> event_indices[start_idx <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>                    type_key <span class="op">=</span> <span class="bu">int</span>(interval_type)</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(start, end <span class="op">+</span> <span class="dv">1</span>):</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>                        velocity <span class="op">=</span> get_abs_velocity(plot_index, idx)</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>                        riemann_distance <span class="op">=</span> get_riemann_dist(plot_index, idx) <span class="op">/</span> (get_times(plot_index, idx) <span class="op">-</span> get_times(plot_index, idx <span class="op">-</span> <span class="dv">1</span>))</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>                        angle <span class="op">=</span> get_velocity_angle_rel(plot_index, idx)</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>                        type_data[type_key][<span class="st">'velocity'</span>].append(velocity)</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>                        type_data[type_key][<span class="st">'riemann'</span>].append(riemann_distance)</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>                        type_data[type_key][<span class="st">'angle'</span>].append(angle)</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">6</span>))</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>    plt.subplot(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> mode <span class="kw">in</span> [<span class="dv">2</span>, <span class="dv">3</span>]:  </span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>        plt.scatter(</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>            type_data[mode][<span class="st">'velocity'</span>],</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>            type_data[mode][<span class="st">'riemann'</span>],</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>            color<span class="op">=</span>colors[mode],</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>            label<span class="op">=</span>labels[mode],</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>            alpha<span class="op">=</span><span class="fl">0.7</span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">"Velocity"</span>)</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">"Riemann Distance"</span>)</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">"Velocity vs Riemann Distance"</span>)</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>    plt.legend()</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>    plt.subplot(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> mode <span class="kw">in</span> [<span class="dv">2</span>, <span class="dv">3</span>]: </span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a>        plt.scatter(</span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a>            type_data[mode][<span class="st">'angle'</span>],</span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a>            type_data[mode][<span class="st">'riemann'</span>],</span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a>            color<span class="op">=</span>colors[mode],</span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a>            label<span class="op">=</span>labels[mode],</span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a>            alpha<span class="op">=</span><span class="fl">0.7</span></span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">"Angle"</span>)</span>
<span id="cb10-53"><a href="#cb10-53" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">"Riemann Distance"</span>)</span>
<span id="cb10-54"><a href="#cb10-54" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">"Angle vs Riemann Distance"</span>)</span>
<span id="cb10-55"><a href="#cb10-55" aria-hidden="true" tabindex="-1"></a>    plt.legend()</span>
<span id="cb10-56"><a href="#cb10-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-57"><a href="#cb10-57" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb10-58"><a href="#cb10-58" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="st">"scatter_two_95.png"</span>)</span>
<span id="cb10-59"><a href="#cb10-59" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="row">
<div class="column" style="width:49%;">
<div id="fig-scat95" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-scat95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/scatter_two_95.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-scat95-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;19: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div>
<div class="column" style="width:49%;">
<div id="fig-scat90" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-scat90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/scatter_two_90.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-scat90-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;20: Confidence level 0.95.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="conformal-mapping" class="level1">
<h1>Conformal mapping</h1>
<p>The conformal mapping is a mathematical transformation that preserves local angles and shapes while potentially altering the size. In this study, we apply conformal maps based on the minimization of area distortion to transform the cell shape into a disk. This transformation allows us to observe protrusions locally, relative to a reference shape obtained after applying the conformal map. The algorithm, as described and implemented in prior works <span class="citation" data-cites="Hu_Zou_Hua_2014">(<a href="#ref-Hu_Zou_Hua_2014" role="doc-biblioref">Hu, Zou, and Hua 2014</a>)</span>.</p>
<p>The example of topographical representation of a cell can be seen below {<a href="#fig-com87" class="quarto-xref">Figure&nbsp;21</a>}:</p>
<div class="columns">
<div class="column" style="width:100%;">
<div id="fig-com87" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-com87-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/combined_cell_87.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-com87-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;21: Cell Shape and Conformal Topographical Representation Over Time.
</figcaption>
</figure>
</div>
</div>
</div>
<p>As discussed in the project proposal, during cell migration, the cytoskeleton undergoes changes that lead to the formation of protrusions on the cell membrane. These protrusions can be detected in two ways:</p>
<ol type="1">
<li><p><strong>Curvature Analysis</strong>: By computing the curvature of the cell shape and identifying its local extrema, which indicate significant deformations of cell shape.</p></li>
<li><p><strong>Conformal Mapping Analysis</strong>: By calculating the difference between the topological cell representation obtained after applying conformal mapping and identifying local maxima, which reveal the locations of protrusions.</p></li>
</ol>
<p>Clement Soubrier adjusted the <strong>unwrap2D</strong> framework <span class="citation" data-cites="Zhou2023.04.12.536640">(<a href="#ref-Zhou2023.04.12.536640" role="doc-biblioref">Zhou et al. 2023</a>)</span> for our specific problem and developed the code to accumulate protrusion statistics using both methods. The results are presented below.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div id="fig-confcurv" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-confcurv-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/curv_conformal.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-confcurv-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;22: Curvature defined protrusion map
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div id="fig-conftop" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-conftop-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/conf_topology.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-conftop-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;23: Topologically defined protrusion map
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-confcurv" class="quarto-xref">Figure&nbsp;22</a> and <a href="#fig-conftop" class="quarto-xref">Figure&nbsp;23</a> demonstrate pie charts showing the difference between the protrusion direction and the direction of cell migration. These figures reveal that protrusions emerge at the front and rear of the cell, a behavior consistently observed in many experiments <span class="citation" data-cites="SenGupta_Parent_Bear_2021">(<a href="#ref-SenGupta_Parent_Bear_2021" role="doc-biblioref">SenGupta, Parent, and Bear 2021</a>)</span>. <a href="#fig-confcurv" class="quarto-xref">Figure&nbsp;22</a> additionally highlights that the protrusion region at the front of the cell is broader than at the rear, which may be associated with the formation of lamellipodia.</p>
<p>This section demonstrated how conformal mapping can be used to investigate cell shape dynamics. However, the analysis conducted here is not limited to the whole dataset examples. Future research could focus on more detailed single-cell analyses, exploring the interplay between protrusion dynamics and cell migration.</p>
<p>Additional examples of conformal mapping application to single cell analysis are provided in (<a href="#sec-someconf" class="quarto-xref">Section&nbsp;6.0.6</a>).</p>
</section>
<section id="conclusion" class="level1">
<h1>Conclusion</h1>
<p>In this project, we conducted an analysis of cell migration and shape dynamics processes. We examined cell trajectories over time, identifying migration modes and transitions between motility regimes using a segmentation and classification framework.</p>
<p>While we observed that Riemann velocities coincided with certain motion switch events, no consistent global correlation was found between Riemann velocity behavior and these transitions. This lack of correlation might be attributed to errors in cell segmentation. Repeating the experiment with a resegmented dataset might enhance accuracy.</p>
<p>Despite this, the accumulated statistical data of Riemann velocities highlights that cell shape variations depend significantly on the migration modes observed. Cell spatial velocities and directional angles also varied during transitions between migration modes.</p>
<p>Conformal mapping, a promising tool applied in this study, reaffirmed classical experimental results on direction of protrusion growth. However, at this stage, we have only minimally explored the use of conformal mapping for investigating specific artifacts of cell migration. In future, it would be valuable to focus on single-cell analysis identifying similar patterns of cell behavior, such as the formation of multiple large protrusions or uniform cell growth. Quantifying the protrusion analysis(counting them, locating, measuring size etc) would be promising as well.</p>
<p>Besides, our research opens the door for combining Riemann velocities analysis with conformal mapping tools. One may detect the protrusion events observed due to conformal mapping and align it with Riemann dynamics.</p>
<p>Thus, based on these conclusions, the following future plans are proposed:</p>
</section>
<section id="future-plans" class="level1">
<h1>Future plans</h1>
<ol type="1">
<li><p><strong>Dataset expansion</strong>: increase the dataset size to observe more characteristic transitions between migration modes and obtain more data for the statistical analysis.</p></li>
<li><p><strong>Resegmentation</strong>: test the existing methods on a resegmented dataset.</p></li>
<li><p><strong>Improve single cell conformal analysis</strong>: looking for consistent cooperative protrusion patterns appear at different event, potentially protrusions distribution.</p></li>
<li><p><strong>Quantifying the protrusion</strong>: develop mathematical tools to describe properties of protrusive behavior, including size, quantity, and shape of individual protrusions.</p></li>
<li><p><strong>Testing another classifier</strong>: since the classifier we employ doesn’t take the superdiffusion mode into consideration, it might be useful to take another one. For example, DL-MSS might help <span class="citation" data-cites="Arts_Smal_Paul_Wyman_Meijering_2019">(<a href="#ref-Arts_Smal_Paul_Wyman_Meijering_2019" role="doc-biblioref">Arts et al. 2019</a>)</span> .</p></li>
<li><p><strong>Combine Riemann Velocity and Conformal Mapping tools</strong>: align Riemann velocity dynamics with protrusion patterns identified using conformal mapping. Moreover, looking at less motile cells may simplify the observation of protrusions.</p></li>
</ol>
</section>
<section id="appendix" class="level1">
<h1>Appendix</h1>
<section id="sec-organization" class="level3">
<h3 class="anchored" data-anchor-id="sec-organization">Dataset organization</h3>
<p>To simplify the analysis, centroid and time data were organized into arrays.</p>
<div id="2c6d3e24" class="cell" data-execution_count="11">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> cell_i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">204</span>):</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    number_of_frames <span class="op">=</span> <span class="bu">sum</span>(os.path.isdir(os.path.join(<span class="ss">f"cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">"</span>, entry)) <span class="cf">for</span> entry <span class="kw">in</span> os.listdir(<span class="ss">f"cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">"</span>))  </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    iter_distance <span class="op">=</span> np.zeros(number_of_frames)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    iter_time <span class="op">=</span> np.zeros(number_of_frames)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    iter_centroid <span class="op">=</span> np.array([np.random.rand(<span class="dv">2</span>) <span class="cf">for</span> _ <span class="kw">in</span> <span class="bu">range</span>(number_of_frames)])</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(number_of_frames):</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>        iter_time[i] <span class="op">=</span> np.load(<span class="ss">f'cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">/frame_</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">/time.npy'</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>        iter_centroid[i] <span class="op">=</span> np.load(<span class="ss">f'cells/cell_</span><span class="sc">{</span>cell_i<span class="sc">}</span><span class="ss">/frame_</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">/centroid.npy'</span>)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    riemann_distances.append(iter_distance)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    times.append(iter_time)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    centroids.append(iter_centroid)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>data_path <span class="op">=</span> <span class="co">########</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(data_path<span class="op">+</span><span class="st">"/times.npy"</span>, <span class="st">'wb'</span>) <span class="im">as</span> f:</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>    np.save(f, np.array(times, dtype<span class="op">=</span><span class="bu">object</span>))</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(data_path<span class="op">+</span><span class="st">"/centroid.npy"</span>, <span class="st">'wb'</span>) <span class="im">as</span> f:</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>    np.save(f, np.array(centroids, dtype<span class="op">=</span><span class="bu">object</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-dcmssprep" class="level3">
<h3 class="anchored" data-anchor-id="sec-dcmssprep">DC-MSS data preparation</h3>
<div id="4efcb40e" class="cell" data-execution_count="12">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.io <span class="im">import</span> savemat</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                                                                            </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>tracks <span class="op">=</span> {}</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, trajectory <span class="kw">in</span> <span class="bu">enumerate</span>(data):</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    n_frames <span class="op">=</span> trajectory.shape[<span class="dv">0</span>]</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    row <span class="op">=</span> np.zeros(n_frames <span class="op">*</span> <span class="dv">8</span>)  </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j, (x, y) <span class="kw">in</span> <span class="bu">enumerate</span>(trajectory):</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>        start_idx <span class="op">=</span> j <span class="op">*</span> <span class="dv">8</span>  </span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>        row[start_idx] <span class="op">=</span> x </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>        row[start_idx <span class="op">+</span> <span class="dv">1</span>] <span class="op">=</span> y  </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    tracks[<span class="ss">f"track_</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">"</span>] <span class="op">=</span> row  </span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>output_path <span class="op">=</span> <span class="st">"trajectory_data.mat"</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>savemat(output_path, {<span class="st">'tracks'</span>: tracks})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>This piece of code converts trajectory data into a .mat file, compatible with MATLAB.</p>
</section>
<section id="sec-align" class="level3">
<h3 class="anchored" data-anchor-id="sec-align">Alignment</h3>
<p>The alignment function (kindly provided by Wanxin Li) ensures proper alignment of cell shapes:</p>
<div id="6c64451a" class="cell" data-execution_count="13">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> align(point, base_point, rescale, rotation, reparameterization, k_sampling_points): <span class="co">#rotation set as False</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Align point and base_point via quotienting out translation, rescaling, rotation and reparameterization</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    total_space <span class="op">=</span> DiscreteCurvesStartingAtOrigin(k_sampling_points<span class="op">=</span>k_sampling_points)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Quotient out translation </span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    point <span class="op">=</span> total_space.projection(point) </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    point <span class="op">=</span> point <span class="op">-</span> gs.mean(point, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    base_point <span class="op">=</span> total_space.projection(base_point)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    base_point <span class="op">=</span> base_point <span class="op">-</span> gs.mean(base_point, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Quotient out rescaling</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rescale:</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>        point <span class="op">=</span> total_space.normalize(point) </span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>        base_point <span class="op">=</span> total_space.normalize(base_point)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Quotient out rotation</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rotation:</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>        point <span class="op">=</span> rotation_align(point, base_point, k_sampling_points)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Quotient out reparameterization</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> reparameterization:</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>        aligner <span class="op">=</span> DynamicProgrammingAligner(total_space)</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>        total_space.fiber_bundle <span class="op">=</span> ReparametrizationBundle(total_space, aligner<span class="op">=</span>aligner)</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>        point <span class="op">=</span> total_space.fiber_bundle.align(point, base_point)</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> point</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-riemann_simple" class="level3">
<h3 class="anchored" data-anchor-id="sec-riemann_simple">Plotting Riemann Velocities of time</h3>
<div id="184a361f" class="cell" data-execution_count="14">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_riemann_cell(plot_index):</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    riemann_distances <span class="op">=</span> []</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    time_data <span class="op">=</span> []</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    num_frames <span class="op">=</span> <span class="bu">len</span>(centr[plot_index<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> frame <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, num_frames):</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>        dist_value <span class="op">=</span> get_riemann_dist(plot_index<span class="op">-</span><span class="dv">1</span>, frame) <span class="op">/</span> (get_times(plot_index<span class="op">-</span><span class="dv">1</span>, frame) <span class="op">-</span> get_times(plot_index<span class="op">-</span><span class="dv">1</span>, frame <span class="op">-</span> <span class="dv">1</span>))</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>        riemann_distances.append(dist_value)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>        time_data.append(get_times(plot_index<span class="op">-</span><span class="dv">1</span>, frame))</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>,<span class="dv">6</span>))</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    plt.plot(time_data, riemann_distances, marker<span class="op">=</span><span class="st">'o'</span>, linestyle<span class="op">=</span><span class="st">'-'</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">"Time"</span>)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">"Riemann Velocities"</span>)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="ss">f"Cell </span><span class="sc">{</span>plot_index<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>    plt.grid(<span class="va">True</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center" style="display: block; margin-left: auto; margin-right: auto;">
<figure class="figure">
<p><img src="img/r.png" class="img-fluid figure-img" style="width:65.0%"></p>
<figcaption>Riemann Velocities graph for cell №87</figcaption>
</figure>
</div>
</section>
<section id="sec-riemann_segmentation" class="level3">
<h3 class="anchored" data-anchor-id="sec-riemann_segmentation">Plotting Riemann Distances with segmentation</h3>
<div id="0e8ea862" class="cell" data-execution_count="15">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> riemann_single_cell_classification(cell_number):</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    cell_index <span class="op">=</span> cell_number <span class="op">-</span> <span class="dv">1</span>  </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    riemann_distances <span class="op">=</span> []</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    time_data <span class="op">=</span> []</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    type_data <span class="op">=</span> defaultdict(<span class="bu">list</span>)  </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="st">'time_events_90.h5'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>        track_i_data <span class="op">=</span> f[<span class="ss">f'/track_</span><span class="sc">{</span>cell_index<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">'</span>][:]</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>        first_three_rows <span class="op">=</span> track_i_data[:<span class="dv">3</span>]</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>        event_indices <span class="op">=</span> first_three_rows[<span class="dv">0</span>, :].astype(<span class="bu">int</span>) <span class="op">-</span> <span class="dv">1</span>  </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>        interval_types <span class="op">=</span> first_three_rows[<span class="dv">2</span>, :]  <span class="co"># </span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    time_data <span class="op">=</span> [times[cell_index][idx] <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="bu">len</span>(times[cell_index]))]  </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    riemann_distances <span class="op">=</span> [</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>        get_riemann_dist(cell_index, idx) <span class="op">/</span> (get_times(cell_index, idx) <span class="op">-</span> get_times(cell_index, idx <span class="op">-</span> <span class="dv">1</span>))</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> idx <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="bu">len</span>(times[cell_index]))</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>    interval_colors <span class="op">=</span> {</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>        <span class="dv">0</span>: <span class="st">"brown"</span>,      </span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>        <span class="dv">1</span>: <span class="st">"blue"</span>,       </span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>        <span class="dv">2</span>: <span class="st">"cyan"</span>,       </span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>        <span class="dv">3</span>: <span class="st">"magenta"</span>,    </span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>        <span class="st">"unclassified"</span>: <span class="st">"black"</span>  </span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> start_idx, interval_type <span class="kw">in</span> <span class="bu">enumerate</span>(interval_types):</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>        start <span class="op">=</span> event_indices[start_idx]</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>        end <span class="op">=</span> event_indices[start_idx <span class="op">+</span> <span class="dv">1</span>] <span class="cf">if</span> start_idx <span class="op">+</span> <span class="dv">1</span> <span class="op">&lt;</span> <span class="bu">len</span>(event_indices) <span class="cf">else</span> <span class="bu">len</span>(time_data) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> start <span class="op">&lt;</span> <span class="bu">len</span>(time_data) <span class="kw">and</span> end <span class="op">&lt;</span> <span class="bu">len</span>(time_data):</span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>            time_segment <span class="op">=</span> time_data[start:end <span class="op">+</span> <span class="dv">1</span>]  </span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>            segment <span class="op">=</span> riemann_distances[start:end <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>            interval_type <span class="op">=</span> <span class="bu">int</span>(interval_type) <span class="cf">if</span> <span class="kw">not</span> np.isnan(interval_type) <span class="cf">else</span> <span class="st">"unclassified"</span></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>            color <span class="op">=</span> interval_colors.get(interval_type, <span class="st">"black"</span>)</span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>            type_data[interval_type].extend(segment)</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>            ax.plot(time_segment, segment, color<span class="op">=</span>color)</span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> matplotlib.lines <span class="im">import</span> Line2D</span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>    legend_elements <span class="op">=</span> [</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"brown"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Immobile"</span>),</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"blue"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Confined Diffusion"</span>),</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"cyan"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Free Diffusion"</span>),</span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"magenta"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Directed Diffusion"</span>),</span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>        Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span><span class="st">"black"</span>, lw<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">"Unclassified"</span>)</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>    ax.legend(handles<span class="op">=</span>legend_elements, loc<span class="op">=</span><span class="st">"upper right"</span>, fontsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>    ax.set_xlabel(<span class="st">"Time"</span>)</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a>    ax.set_ylabel(<span class="st">"Riemann velocity"</span>)</span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="ss">f"Cell </span><span class="sc">{</span>cell_number<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a>    ax.tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'major'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="ss">f"riemann_single_cell_</span><span class="sc">{</span>cell_number<span class="sc">}</span><span class="ss">_classification_90.png"</span>)</span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-someconf" class="level3">
<h3 class="anchored" data-anchor-id="sec-someconf">Additional examples: cell №87 topographical representations over migration modes transitions</h3>
<p>This section doesn’t include any sufficient results to report in the blog, but there we tried to collect some evidence of a single cell shape artifacts which emerge during migration modes transitions.</p>
<div class="columns">
<div class="column" style="width:100%;">
<div id="fig-app1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-app1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/combined_cell_87_1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-app1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;24: t = 51
</figcaption>
</figure>
</div>
</div>
</div>
<p>From the <a href="#fig-app1" class="quarto-xref">Figure&nbsp;24</a>, we observe that when the cell transitions from directed migration to a free or confined diffusion mode, a new noticeable protrusion is formed, and the overall structure of the cell membrane undergoes significant multilateral changes.</p>
<div class="columns">
<div class="column" style="width:100%;">
<div id="fig-app2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-app2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/combined_cell_87_2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-app2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;25: t = 85
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-app2" class="quarto-xref">Figure&nbsp;25</a> shows how the cell increases its volume almost uniformly across the entire membrane curve.</p>
<div class="columns">
<div class="column" style="width:100%;">
<div id="fig-app3" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-app3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/combined_cell_87_3.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-app3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;26: t = 230
</figcaption>
</figure>
</div>
</div>
</div>
<p>Topological analysis <a href="#fig-app3" class="quarto-xref">Figure&nbsp;26</a> reveals how the cell transforms two protrusions into a single one while decreasing its volume.</p>
<p>Since these conclusions are superficial and far-fetched, the farther exploration of single-cell topological representations might lead to a significant advance of this research.</p>
</section>
</section>
<section id="references" class="level1">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Arts_Smal_Paul_Wyman_Meijering_2019" class="csl-entry" role="listitem">
Arts, Marloes, Ihor Smal, Maarten W. Paul, Claire Wyman, and Erik Meijering. 2019. <span>“Particle Mobility Analysis Using Deep Learning and the Moment Scaling Spectrum.”</span> <em>Nature News</em>. Nature Publishing Group. <a href="https://www.nature.com/articles/s41598-019-53663-8">https://www.nature.com/articles/s41598-019-53663-8</a>.
</div>
<div id="ref-Modes" class="csl-entry" role="listitem">
Dieterich, Peter, Rainer Klages, Roland Preuss, and Albrecht Schwab. 2008. <span>“Anomalous Dynamics of Cell Migration.”</span> <em>Proceedings of the National Academy of Sciences</em> 105 (2): 459–63. <a href="https://doi.org/10.1073/pnas.0707603105">https://doi.org/10.1073/pnas.0707603105</a>.
</div>
<div id="ref-Hu_Zou_Hua_2014" class="csl-entry" role="listitem">
Hu, Jiaxi, Guangyu Jeff Zou, and Jing Hua. 2014. <span>“Volume-Preserving Mapping and Registration for Collective Data Visualization.”</span> <em>IEEE Transactions on Visualization and Computer Graphics</em> 20 (12): 2664–73. <a href="https://doi.org/10.1109/tvcg.2014.2346457">https://doi.org/10.1109/tvcg.2014.2346457</a>.
</div>
<div id="ref-picture" class="csl-entry" role="listitem">
Kapanidis, Achillefs, Stephan Uphoff, and Mathew Stracy. 2018. <span>“Understanding Protein Mobility in Bacteria by Tracking Single Molecules.”</span> <em>Journal of Molecular Biology</em> 430 (May). <a href="https://doi.org/10.1016/j.jmb.2018.05.002">https://doi.org/10.1016/j.jmb.2018.05.002</a>.
</div>
<div id="ref-10.1007/978-3-031-38299-4_60" class="csl-entry" role="listitem">
Li, Wanxin, Ashok Prasad, Nina Miolane, and Khanh Dao Duc. 2023. <span>“Using a&nbsp;Riemannian Elastic Metric for&nbsp;Statistical Analysis of&nbsp;Tumor Cell Shape Heterogeneity.”</span> In <em>Geometric Science of Information</em>, edited by Frank Nielsen and Frédéric Barbaresco, 583–92. Cham: Springer Nature Switzerland.
</div>
<div id="ref-MSD" class="csl-entry" role="listitem">
Rosen, Christopher P. AND Dallon, Mary Ellen AND Grant. 2021. <span>“Mean Square Displacement for a Discrete Centroid Model of Cell Motion.”</span> <em>PLOS ONE</em> 16 (12): 1–19. <a href="https://doi.org/10.1371/journal.pone.0261021">https://doi.org/10.1371/journal.pone.0261021</a>.
</div>
<div id="ref-SenGupta_Parent_Bear_2021" class="csl-entry" role="listitem">
SenGupta, Shuvasree, Carole A. Parent, and James E. Bear. 2021. <span>“The Principles of Directed Cell Migration.”</span> <em>Nature Reviews Molecular Cell Biology</em> 22 (8): 529–47. <a href="https://doi.org/10.1038/s41580-021-00366-6">https://doi.org/10.1038/s41580-021-00366-6</a>.
</div>
<div id="ref-VEGA20181018" class="csl-entry" role="listitem">
Vega, Anthony R., Spencer A. Freeman, Sergio Grinstein, and Khuloud Jaqaman. 2018. <span>“Multistep Track Segmentation and Motion Classification for Transient Mobility Analysis.”</span> <em>Biophysical Journal</em> 114 (5): 1018–25. https://doi.org/<a href="https://doi.org/10.1016/j.bpj.2018.01.012">https://doi.org/10.1016/j.bpj.2018.01.012</a>.
</div>
<div id="ref-wikipedia_anomalous_diffusion" class="csl-entry" role="listitem">
Wikipedia. n.d. <span>“Anomalous Diffusion.”</span> <a href="https://en.wikipedia.org/wiki/Anomalous_diffusion">https://en.wikipedia.org/wiki/Anomalous_diffusion</a>.
</div>
<div id="ref-Zhou2023.04.12.536640" class="csl-entry" role="listitem">
Zhou, Felix Y., Andrew Weems, Gabriel M. Gihana, Bingying Chen, Bo-Jui Chang, Meghan Driscoll, and Gaudenz Danuser. 2023. <span>“Surface-Guided Computing to Analyze Subcellular Morphology and Membrane-Associated Signals in 3D.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2023.04.12.536640">https://doi.org/10.1101/2023.04.12.536640</a>.
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/bioshape-analysis\.github\.io\/blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>